
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Practice: Introduction to gene expression (RNAseq) analysis &#8212; Plants &amp; Python</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe,.cell"
        const thebe_selector_input = "pre,.cell_input div.highlight"
        const thebe_selector_output = ".output,.cell_output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Welcome / Bienvenid@" href="../R_11_VARIANT_DISCOVERY/0_Overview.html" />
    <link rel="prev" title="10.0 A short introduction to short read analysis" href="0_Lessons/10.0_RNAseq_analysis.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/plants_python_logo.jpg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Plants & Python</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../00_Opening_page.html">
   Welcome to / Bienvenid@ a Plants &amp; Python
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Preface
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../00_PREFACE/01_Overview.html">
   Welcome : Bienvenid@
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/01_Welcome_Bienvenido/00_Welcome.html">
     Welcome to Plants &amp; Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/01_Welcome_Bienvenido/01_Bienvenido.html">
     Bienvenid@ a Plants &amp; Python
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../00_PREFACE/Introduction_Illustration.html">
   Introduction Illustration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../00_PREFACE/Ilustracion_de_introduccion.html">
   Ilustración de introducción
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../00_PREFACE/02_Overview.html">
   Getting Started : Empezando
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/00_Installing_Anaconda.html">
     1. Installing Anaconda and using Jupyter notebooks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/01_Instalar_Anaconda.html">
     1. Instalar Anaconda y usar Jupyter notebooks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/02_Coding_cells.html">
     2. Coding cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/03_Celdas_de_codigo.html">
     2. Celdas de código
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/04_Markdown_cells.html">
     3. Markdown cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/02_Getting_started_Empezando/05_Celdas_de_markdown.html">
     3. Celdas de markdown
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../00_PREFACE/03_Overview.html">
   Using : Usando Jupyter Book
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/03_Using_Usando_Jupyter_Book/How_to_use_this_Jupyter_Book.html">
     How to use this Jupyter Book
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00_PREFACE/03_Using_Usando_Jupyter_Book/Como_usar_este_Jupyter_Book.html">
     Cómo usar este Jupyter Book
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  1. Variables, lists, &amp; indexing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/Cover_image_1.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.0_Variables.html">
     1.0 Variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.1_Functions.html">
     1.1 Functions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.2_Indexing.html">
     1.2 Indexing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.3_Lists.html">
     1.3 Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.4_Indexing_List_Elements.html">
     1.4 Indexing List Elements
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.5_Modifying_Lists.html">
     1.5 Modifying Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.6_List_of_Lists.html">
     1.6 List of Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.7_Slicing.html">
     1.7 Slicing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.8_Data_Types.html">
     1.8 Data Types
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../A_1_VARIABLES_LISTS_AND_INDEXING/2_Plants_And_Climate.html">
   Plants &amp; Climate
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  1. Variables, listas, &amp; INDEXACIÓN
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/Ilustracion_de_portada_1.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.0_Variables.html">
     1.0 Variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.1_Funciones.html">
     1.1 Funciones
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.2_Indexacion.html">
     1.2 Indexación
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.3_Listas.html">
     1.3 Listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.4_Indexando_los_elementos_de_una_lista.html">
     1.4 Indexando los elementos de una lista
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.5_Modificando_listas.html">
     1.5 Modificando listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.6_Lista_de_listas.html">
     1.6 Lista de listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.7_Rebanar.html">
     1.7 Rebanar
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.8_Tipos_de_datos.html">
     1.8 Tipos de datos
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../B_1_VARIABLES_LISTAS_Y_INDEXACION/2_Las_plantas_y_el_clima.html">
   Las plantas y el clima
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  2. Visualizing Data with Matplotlib
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/Cover_image_2.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.0_Modules.html">
     2.0 Modules
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.1_Plots.html">
     2.1 Plots
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.2_Shape_Size_and_Color.html">
     2.2 Shape, Size, and Color
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.3_Legends.html">
     2.3 Legends
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.4_Subplots.html">
     2.4 Subplots
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/2_Leaf_And_Sunflower_Shapes.html">
   Leaf &amp; Sunflower Shapes
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  2. VISUALIZACIÓN DE DATOS CON MATPLOTLIB
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/Ilustracion_de_portada_2.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.0_Modulos.html">
     2.0 Módulos
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.1_Graficos.html">
     2.1 Gráficos
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.2_Forma_tamano_y_color.html">
     2.2 Forma, tamaño y color
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.3_Leyendas.html">
     2.3 Leyendas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.4_Subgraficos.html">
     2.4 Subgráficos
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/2_Las_formas_de_las_hojas_y_girasoles.html">
   Las formas de las hojas y girasoles
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  3. Loops &amp; the Golden Angle
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/Cover_image_3.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.0_Range_Function.html">
     3.0 Range Function
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.1_For_Loops.html">
     3.1 For Loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.2_Counters.html">
     3.2 Counters
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.3_Using_Lists_and_Loops.html">
     3.3 Using Lists and Loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.4_Calculating_Fibonacci.html">
     3.4 Calculating Fibonacci
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.5_While_Loops.html">
     3.5 While Loops
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/2_Modeling_Sunflowers.html">
   Modeling Sunflowers
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  3. LOOPS &amp; EL ÁNGULO ÁUREO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/Ilustracion_de_portada_3.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.0_La_funcion_range.html">
     3.0 La función range
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.1_Loops_for.html">
     3.1 Loops for
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.2_Contadores.html">
     3.2 Contadores
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.3_Usando_listas_y_loops.html">
     3.3 Usando listas y loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.4_Calcular_Fibonacci.html">
     3.4 Calcular Fibonacci
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.5_Loops_while.html">
     3.5 Loops while
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../F_3_LOOPS_Y_EL_ANGULO_DORADO/2_Como_construir_un_girasol.html">
   Cómo construir un girasol
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  4. Data Analysis with Pandas
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/Cover_image_4.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.0_Reading_in_Data.html">
     4.0 Reading in Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.1_Head_and_Tail.html">
     4.1 Head and Tail
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.2_Selecting_Columns.html">
     4.2 Selecting Columns
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.3_Descriptive_Statistics.html">
     4.3 Descriptive Statistics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.4_Masks.html">
     4.4 Masks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.5_Using_iloc.html">
     4.5 Using iloc
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.6_Seaborn.html">
     4.6 Seaborn
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../G_4_DATA_ANALYSIS_WITH_PANDAS/2_Grapes_in_a_Warming_World.html">
   Grapes in a Warming World
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  4. Análisis de datos con pandas
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/Ilustracion_de_portada_4.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.0_Leyendo_en_data.html">
     4.0 Leyendo en data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.1_Head_y_tail.html">
     4.1 Head y tail
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.2_Seleccionando_columnas.html">
     4.2 Seleccionando columnas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.3_Estadisticas_descriptivas.html">
     4.3 Estadísticas descriptivas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.4_Enmascaramiento.html">
     4.4 Enmascaramiento
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.5_Usando_iloc.html">
     4.5 Usando iloc
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.6_Seaborn.html">
     4.6 Seaborn
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../H_4_ANALISIS_DE_DATOS_CON_PANDAS/2_Vides_en_un_mundo_que_se_calienta.html">
   Vides en un mundo que se calienta
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  5. Unix Command Line
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/Cover_image_5.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.0_Introduction_to_Command_Line.html">
     <p style="color:#2C6A7F">
      5.0 Introduction to Command Line
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.1_TheVirtual_Terminal.html">
     <p style="color:#2C6A7F">
      5.1 Virtual Terminal
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.2_Exploring_Folders.html">
     <p style="color:#2C6A7F">
      5.2 Exploring Folders
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.3_Working_with_Files.html">
     <p style="color:#2C6A7F">
      5.3 Working with Files
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.4_Options_and_Arguments.html">
     <p style="color:#2C6A7F">
      5.4 Options and Arguments
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/0_Lessons/5.5_Copy_Move_and_Remove_Files.html">
     <p style="color:#2C6A7F">
      5.5 Copy, Move &amp; Remove Files
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../I_5_UNIX_COMMAND_LINE/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  5. LINEA DE COMANDOS DE UNIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/Ilustracion_de_portada_5.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.0_IntroduccionLineaDeComandos.html">
     <p style="color:#2C6A7F">
      5.0 Introducción a la línea de comandos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.1_LaTerminal_Virtual.html">
     <p style="color:#2C6A7F">
      5.1 Terminal Virtual
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.2_ExplorandoDirectorios.html">
     <p style="color:#2C6A7F">
      5.2 Explorando directorios
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.3_Trabajando_con_archivos.html">
     <p style="color:#2C6A7F">
      5.3 Trabajando con archivos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.4_Opciones_y_Argumentos.html">
     <p style="color:#2C6A7F">
      5.4 Opciones y argumentos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.5_Copiar_Mover_Eliminar_Archivos.html">
     <p style="color:#2C6A7F">
      5.5 Copiar, mover y eliminar archivos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../J_5_LINEA_DE_COMANDOS_DE_UNIX/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  6. Unix Commands For Data Mining
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
  <label for="toctree-checkbox-14">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.0_Edit_Sort_Cut_grep.html">
     <p style="color:#2C6A7F">
      6.0 EDIT, SORT, CUT &amp; GREP
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.1_Word_Count_Pipelines_Permissions.html">
     <p style="color:#2C6A7F">
      6.1 Word Count, Pipelines &amp; Permissions
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.2_Biological_Data_Formats.html">
     <p style="color:#2C6A7F">
      6.2 Biological Data Formats
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../K_6_UNIX_COMMANDS_FOR_DATA_MINING/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  6. MINERIA DE DATOS EN LA LINEA DE COMANDOS DE UNIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
  <label for="toctree-checkbox-15">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.0_Editar_Ordentar_Cortar_grep.html">
     <p style="color:#2C6A7F">
      6.0 Editar, ordenar, cortar y grep
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.1_Contar_Palabras_Tuberias_Permisos.html">
     <p style="color:#2C6A7F">
      6.1 Contar palabras, tuberías &amp; permisos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.2_Formatos_Biologicos.html">
     <p style="color:#2C6A7F">
      6.2 Formatos de Datos Biológicos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  7. SCRIPTING WITH BASH
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../M_7_SCRIPTING_WITH_BASH/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
  <label for="toctree-checkbox-16">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.0_WorkingRemotely.html">
     <p style="color:#2C6A7F">
      7.0 Working Remotely
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.1_Scripting_with_bash.html">
     <p style="color:#2C6A7F">
      7.1 Scripting with Bash
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.2_Summary_of_Commands.html">
     <p style="color:#2C6A7F">
      7.2 Summary of Commands
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../M_7_SCRIPTING_WITH_BASH/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  7. PROGRAMANDO CON BASH
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../N_7_PROGRAMANDO_CON_BASH/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
  <label for="toctree-checkbox-17">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.0_Trabajando_de_manera_remota.html">
     <p style="color:#2C6A7F">
      7.0 Trabajando de manera remota
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.1_Programando_con_Bash.html">
     <p style="color:#2C6A7F">
      7.1 Programando con Bash
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.2_Resumen_de_comandos.html">
     <p style="color:#2C6A7F">
      7.2 Resumen de comandos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../N_7_PROGRAMANDO_CON_BASH/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  8. GENOMICS AND BIOPYTHON
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../O_8_GENOMICS_AND_BIOPYTHON/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
  <label for="toctree-checkbox-18">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../O_8_GENOMICS_AND_BIOPYTHON/0_Lessons/8.0_Intro_to_Genomics_and_Biopython.html">
     8.0 Introduction to plant genomics and working with sequences in Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../O_8_GENOMICS_AND_BIOPYTHON/0_Lessons/8.1_HPCC.html">
     Lesson 8.6 Using the high performance computing cluster (HPCC) at MSU
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../O_8_GENOMICS_AND_BIOPYTHON/1_Practice.html">
   Practice: How to Build a Genome
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  9. COMPARATIVE GENOMICS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../P_9_COMPARATIVE_GENOMICS/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
  <label for="toctree-checkbox-19">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../P_9_COMPARATIVE_GENOMICS/0_Lessons/9.0_Comp_Genomics.html">
     9.0 Genome visualization and comparative genomics
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../P_9_COMPARATIVE_GENOMICS/1_Practice.html">
   Practice: Genome visualization and comparative genomics
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  10. RNASEQ ANALYSIS
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
  <label for="toctree-checkbox-20">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="0_Lessons/10.0_RNAseq_analysis.html">
     10.0 A short introduction to short read analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Practice: Introduction to gene expression (RNAseq) analysis
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  11. VARIANT DISCOVERY
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../R_11_VARIANT_DISCOVERY/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
  <label for="toctree-checkbox-21">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../R_11_VARIANT_DISCOVERY/0_Lessons/11.0_Variant_Discovery.html">
     11.0 Variant discovery
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../R_11_VARIANT_DISCOVERY/1_Practice.html">
   Practice: Variant discovery, genetic diversity, and population genetics
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/Q_10_RNASEQ_ANALYSIS/1_Practice.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/PlantsAndPython/PlantsAndPython"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/PlantsAndPython/PlantsAndPython/issues/new?title=Issue%20on%20page%20%2FQ_10_RNASEQ_ANALYSIS/1_Practice.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/PlantsAndPython/PlantsAndPython/master?urlpath=tree/Q_10_RNASEQ_ANALYSIS/1_Practice.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/PlantsAndPython/PlantsAndPython/blob/master/Q_10_RNASEQ_ANALYSIS/1_Practice.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        <button type="button" class="btn btn-secondary topbarbtn"
            onclick="initThebeSBT()" title="Launch Thebe" data-toggle="tooltip" data-placement="left"><i
                class="fas fa-play"></i><span style="margin-left: .4em;">Live Code</span></button>
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Practice: Introduction to gene expression (RNAseq) analysis
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#read-alignment">
   Read alignment
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#question-what-is-the-read-mapping-percentage-of-this-sample-do-you-think-this-is-good-what-factors-could-be-driving-this-mapping-rate">
     Question: What is the read mapping percentage of this sample? Do you think this is good? What factors could be driving this mapping rate?
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-import-and-exploration">
   Data import and exploration
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#challenge-problem-rnaseq-pipeline">
   Challenge problem: RNAseq pipeline
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <p><img alt="plants_python_logo.jpg" src="../_images/plants_python_logo.jpg" /></p>
<div class="section" id="practice-introduction-to-gene-expression-rnaseq-analysis">
<h1>Practice: Introduction to gene expression (RNAseq) analysis<a class="headerlink" href="#practice-introduction-to-gene-expression-rnaseq-analysis" title="Permalink to this headline">¶</a></h1>
<p>In this activity, we will learn about different read alignment and quantification programs, analyze gene expression data from maize, interpret the outputs, and run some quality control analyses on a matrix of RNAseq expression values.</p>
</div>
<div class="section" id="read-alignment">
<h1>Read alignment<a class="headerlink" href="#read-alignment" title="Permalink to this headline">¶</a></h1>
<p>There are many ways to analyze gene expression data, starting from how the reads are aligned/mapped against the genome, how we identify genes that are differentially expressed, and the types of downstream analyses we do with the data. In this notebook we will focus on (1) how to quantify gene expression data, (2) how to manipulate an expression-based dataframe, (3) and how we can assess the quality of the data prior to downstream analyses. We will work on analyzing the expression data in the next notebook. <br>
In our raw fastq file, we have fragments of sequenced mRNAs and we want to know which genes they came from. We can do this by <strong>aligning</strong> or <strong>mapping</strong> the reads against the genome to figure out the exact base pair overlap of the reads and the genome, or we can <strong>pseudo-align</strong> or <strong>quasi-map</strong> the reads to figure out the relative position of where the reads align. These two approaches produce <em>largely</em> the same results, but each has unique advantages and disadvantages. <br>
Allignment based algorithms are more computationally intense because after they determine which region in the genome the reads map to, run a local alignment, and record the exact coordinates and any differences between the genome and read. Alignemnt based programs produces a really large <code class="docutils literal notranslate"><span class="pre">.sam</span></code> file which contains the read information and where it aligns. A <code class="docutils literal notranslate"><span class="pre">.sam</span></code> file is essential for doing variant analyses and for quantifying alternative splicing, but it is not needed to simply count reads per gene.  Alignment of RNAseq data is also challenging because the genome sequence contains both the introns and exons and RNAseq reads contain only exons (introns are spliced out), so a single read can map to several distant regions in the genome. There are several leading alignment based programs for RNAseq including HISAT2 and STAR (among many others, don’t &#64; me!): <br>
HISAT2: <br>
<a class="reference external" href="http://ccb.jhu.edu/software/hisat2/index.shtml">http://ccb.jhu.edu/software/hisat2/index.shtml</a>
STAR: <br>
<a class="reference external" href="https://github.com/alexdobin/STAR">https://github.com/alexdobin/STAR</a>
<br>
Pseudoalignment or quasi-mapping based programs work a little differently. Instead of calculating the optimal alignment and exact overlap between a read and the genome, these programs use a hash table to figure out which transcript the read is from without doing the computationally intensive alignment. Pseudoaligners use the gene models (or transcripts) as a reference for alignment instead of the genome, so they don’t run into problems associated with splicing. Instead of producing an alignment file (<code class="docutils literal notranslate"><span class="pre">.sam</span></code>), pseudoaligners output a table with counts for each gene. Pseudoaligners run <em>ultra fast</em> and can process a stardard RNAseq sample in a few minutes (compared to hours for true aligners). There are a few good pseudoaligners and the two most popular are kallisto and Salmon: <br>
kallisto: <br>
<a class="reference external" href="https://pachterlab.github.io/kallisto/">https://pachterlab.github.io/kallisto/</a>
<br>
Salmon: <br>
<a class="reference external" href="https://combine-lab.github.io/salmon/">https://combine-lab.github.io/salmon/</a> <br></p>
<p>We will use Salmon in the course, but other programs might be better for your experimental system. In this lession we will be using RNAseq data from a few tissues in maize. You can also run the analyses in this notebook on the data you downloaded and trimmed from the SRA in the pre-class lesson.</p>
<p><img alt="salmon.png" src="../_images/salmon.png" /></p>
<p>To run Salmon we need an indexed file of transcripts (to map the reads against) and either unpaired or paired Illumina .fastq files. Salmon is installed on HPCC, and we can run it by loading the Salmon module: <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$module load Salmon/1.2.1
</pre></div>
</div>
<br>
First, we need to build our index. The index is a structure that Salmon uses to quasi-map RNA-seq reads during quantification. The index only needs to be constructed once per transcriptome/species, and it can then be reused to quantify many experiments. If you don't have a set of transcripts (transcriptome) you need to download it for your favorite genome. We can get this from several places including ensembl: 
http://ensemblgenomes.org/info/genomes?division=4
We can use the `wget` command on HPCC to download a file directly from a link. Note: The file will be downloaded automatically in your present working directory, Make sure you are running this in your home directory or scratch <br>  
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$wget http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-51/plants/fasta/zea_mays/cdna/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.cdna.all.fa.gz
</pre></div>
</div>
<br>
Here, we are downloading the set of transcripts from maize in fasta format, but you can run Salmon on your favorite species using data from the SRA (or your research project), and using the .cds or .transcript file from your favorite species.  
We can then use the `index` command of Salmon to build our index for mapping, where the `-t` flag specifies the transcript file (which we just downloaded) and `-i` specifies the directory and prefix for the index file(s) you will be building: <br>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$salmon index -t Zea_mays.Zm-B73-REFERENCE-NAM-5.0.cdna.all.fa.gz -i Zea_mays_Salmon_index
</pre></div>
</div>
<p>After we have an index, we can run the <code class="docutils literal notranslate"><span class="pre">quant</span></code> function to quantify expression. <br>
I have downloaded ~10 files from the SRA corresponding to maize seedling gene expression data. The data can be found under <code class="docutils literal notranslate"><span class="pre">/mnt/research/REU_Plant_Genome_2021/RNAseq/data/</span></code>. We will srart by quantifying expression for one of these samples, but if you are interested, you can quantify expression from all of them (see an example batch submission script below). <br> <br>
To run <code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">quant</span></code>, we will set a few flags as follows. <code class="docutils literal notranslate"><span class="pre">-i</span></code> sets the index <code class="docutils literal notranslate"><span class="pre">-r</span></code> sets the path to the fastq file (if the data is unpaired) <code class="docutils literal notranslate"><span class="pre">-1</span></code> and <code class="docutils literal notranslate"><span class="pre">-2</span></code> set the path for read 1 and read 2 for paired end data. <code class="docutils literal notranslate"><span class="pre">-p</span></code> sets the number of threads to use and <code class="docutils literal notranslate"><span class="pre">-o</span></code> sets the output directory: <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$salmon quant -i Zea_mays_Salmon_index/ -l A -r /mnt/research/HRT841_F21/RNAseq/data/SRR940276.trimmed.fastq -p 6 -o SRR940276.quant
</pre></div>
</div>
<p>You can run this on a developmental node and it should run in a few minutes. We can enter the output directory to find the files we need. The files we care about are the <code class="docutils literal notranslate"><span class="pre">/log/salmon_quant.log</span></code> and <code class="docutils literal notranslate"><span class="pre">quant.sf</span></code> files. The log file contains all the info and metrics from the run and the <code class="docutils literal notranslate"><span class="pre">.sf</span></code> file contains the read counts for each gene. Read counts are important, but these can be misleading because longer genes will be fragmented into more pieces per single mRNA molecule, and might appear to have higher expression. Two genes with similar expression levels could have different read counts if one gene is significanlty larger. The numer of reads per gene (transcript) is also dependent on the total number of reads sequenced for each library.  We can normalize the data by the number of reads per library and the lengths of transcripts to create a metric of transcripts per million (TPM). This will allow us to compare the expression pattern of the same gene across multiple tissues/timepoints.</p>
<div class="section" id="question-what-is-the-read-mapping-percentage-of-this-sample-do-you-think-this-is-good-what-factors-could-be-driving-this-mapping-rate">
<h2>Question: What is the read mapping percentage of this sample? Do you think this is good? What factors could be driving this mapping rate?<a class="headerlink" href="#question-what-is-the-read-mapping-percentage-of-this-sample-do-you-think-this-is-good-what-factors-could-be-driving-this-mapping-rate" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Answer. Hint, the mapping rate is found in the salmon_quant.log file in your salmon output directory. </span>
</pre></div>
</div>
</div>
</div>
<p>We can run individual <code class="docutils literal notranslate"><span class="pre">quant</span></code> commands for each file separately, but this is tedious, especially if we have thousands of RNAseq files. We can use a shell script to run them all automatically. Below is a shell script I wrote to do this. <strong>Note: you will need to change all of the directories here if you choose to test this shell script</strong> <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#!/bin/bash

module load salmon

cd /mnt/research/VanBuren_Lab/Bob/Teaching/RNAseq/

for i in /mnt/research/HRT841_F21/RNAseq/data/*.fastq
do

unpaired=$(echo ${i})
Name=$(echo ${i} | sed &#39;s/\/mnt\/research\/HRT841_f20\/RNAseq\/data\///&#39;| sed &#39;s/.fastq//&#39;)
echo ${i}
# Run salmon command followed by done to end the loop
salmon quant -l a -p 6 -r ${unpaired} -i Zea_mays_Salmon_index/ -o ${Name}_quant
echo ${Name} &amp; done
</pre></div>
</div>
<p>We can follow the logic of shell scripts based on what we know about variables and loops in python. Here, we are running a <code class="docutils literal notranslate"><span class="pre">for</span></code> loop on all of the fastq files within a directory. Shell loops require three elements: for, do, and done. We will set two variables, one to retrieve the reads and one to set the output prefix. Instead of print, we will use echo. For the <code class="docutils literal notranslate"><span class="pre">Name</span></code> variable, echo will print to full path of the file, and we can use <code class="docutils literal notranslate"><span class="pre">sed</span></code> to remove all of the charactors we don’t want. <br>
You can save this shell script with the file extension <code class="docutils literal notranslate"><span class="pre">.sh</span></code> and run it on a developmental node (or submit it to the queue via Slurm): <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$bash salmon_quant_script.sh
</pre></div>
</div>
<p>Our shell sript will process all of the files individually, but we want all of the data in the same dataframe. We could write a script to parse each separate dataframe, rename the columns and merge them, but luckily for us, the Salmon function <code class="docutils literal notranslate"><span class="pre">quantmerge</span></code>exists to do this:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> salmon quantmerge --quants *_quant -o merged TPM
</pre></div>
</div>
<p>We need to specify the directories that we want to merge using the <code class="docutils literal notranslate"><span class="pre">--quants</span></code> flag. Using <code class="docutils literal notranslate"><span class="pre">*</span></code> matches zero or more random charactors and will include any directories that end in “quant” <code class="docutils literal notranslate"><span class="pre">-o</span></code> specifies the prefix of the merged TPM file.</p>
<p><strong>Note:</strong> You can run this shell script on the maize data if you want, but you can also run it on the data you downloaded from the SRA or your own sequencing data.</p>
</div>
</div>
<div class="section" id="data-import-and-exploration">
<h1>Data import and exploration<a class="headerlink" href="#data-import-and-exploration" title="Permalink to this headline">¶</a></h1>
<p>Now that we have a matrix of TPMs for a couple samples, we are ready to analyze the data for differential gene expression, co-expression networks, and other downstream analyses. But first, we need to import the data and we should test for variance between our replicates. You could use the maize data analyzed above or one of the three dataframes I provided on discord. If the data was collected carefully, we would expect a low amount of variance between the replicates, but a lot of variation between tissue types. If the data was collected haphazardly or mishandled, we would expect a high degree of distance between replicates in the PCA. If the replicates have a high amount of variance this will effect the downstream analyses. We can explore the RNAseq data using principal component analysis (PCA).
<br>
First, we need to import out table of TPMs as a dataframe. <strong>Note:</strong> You need to download an example dataframe file from Discord or from HPCC!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">expr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;FILE_NAME&#39;</span><span class="p">)</span>
<span class="n">expr_df</span> <span class="o">=</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">expr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">expr_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="n">expr_df</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="n">ipykernel_30933</span><span class="o">/</span><span class="mf">2169730351.</span><span class="n">py</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> 
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">expr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;FILE_NAME&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> <span class="n">expr_df</span> <span class="o">=</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">expr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="g g-Whitespace">      </span><span class="mi">6</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/util/_decorators.py</span> in <span class="ni">wrapper</span><span class="nt">(*args, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">309</span>                     <span class="n">stacklevel</span><span class="o">=</span><span class="n">stacklevel</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">310</span>                 <span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">311</span>             <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">312</span> 
<span class="g g-Whitespace">    </span><span class="mi">313</span>         <span class="k">return</span> <span class="n">wrapper</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">read_csv</span><span class="nt">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">584</span>     <span class="n">kwds</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds_defaults</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">585</span> 
<span class="ne">--&gt; </span><span class="mi">586</span>     <span class="k">return</span> <span class="n">_read</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">587</span> 
<span class="g g-Whitespace">    </span><span class="mi">588</span> 

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">_read</span><span class="nt">(filepath_or_buffer, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">480</span> 
<span class="g g-Whitespace">    </span><span class="mi">481</span>     <span class="c1"># Create the parser.</span>
<span class="ne">--&gt; </span><span class="mi">482</span>     <span class="n">parser</span> <span class="o">=</span> <span class="n">TextFileReader</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">483</span> 
<span class="g g-Whitespace">    </span><span class="mi">484</span>     <span class="k">if</span> <span class="n">chunksize</span> <span class="ow">or</span> <span class="n">iterator</span><span class="p">:</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">__init__</span><span class="nt">(self, f, engine, **kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">809</span>             <span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwds</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">    </span><span class="mi">810</span> 
<span class="ne">--&gt; </span><span class="mi">811</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_engine</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_engine</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">engine</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">812</span> 
<span class="g g-Whitespace">    </span><span class="mi">813</span>     <span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">_make_engine</span><span class="nt">(self, engine)</span>
<span class="g g-Whitespace">   </span><span class="mi">1038</span>             <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1039</span>         <span class="c1"># error: Too many arguments for &quot;ParserBase&quot;</span>
<span class="ne">-&gt; </span><span class="mi">1040</span>         <span class="k">return</span> <span class="n">mapping</span><span class="p">[</span><span class="n">engine</span><span class="p">](</span><span class="bp">self</span><span class="o">.</span><span class="n">f</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">)</span>  <span class="c1"># type: ignore[call-arg]</span>
<span class="g g-Whitespace">   </span><span class="mi">1041</span> 
<span class="g g-Whitespace">   </span><span class="mi">1042</span>     <span class="k">def</span> <span class="nf">_failover_to_python</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/c_parser_wrapper.py</span> in <span class="ni">__init__</span><span class="nt">(self, src, **kwds)</span>
<span class="g g-Whitespace">     </span><span class="mi">49</span> 
<span class="g g-Whitespace">     </span><span class="mi">50</span>         <span class="c1"># open handles</span>
<span class="ne">---&gt; </span><span class="mi">51</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_open_handles</span><span class="p">(</span><span class="n">src</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">52</span>         <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
<span class="g g-Whitespace">     </span><span class="mi">53</span> 

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/base_parser.py</span> in <span class="ni">_open_handles</span><span class="nt">(self, src, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">220</span>         <span class="n">Let</span> <span class="n">the</span> <span class="n">readers</span> <span class="nb">open</span> <span class="n">IOHandles</span> <span class="n">after</span> <span class="n">they</span> <span class="n">are</span> <span class="n">done</span> <span class="k">with</span> <span class="n">their</span> <span class="n">potential</span> <span class="n">raises</span><span class="o">.</span>
<span class="g g-Whitespace">    </span><span class="mi">221</span>         <span class="s2">&quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">222</span><span class="s2">         self.handles = get_handle(</span>
<span class="g g-Whitespace">    </span><span class="mi">223</span><span class="s2">             src,</span>
<span class="g g-Whitespace">    </span><span class="mi">224</span><span class="s2">             &quot;r&quot;,</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/common.py</span> in <span class="ni">get_handle</span><span class="nt">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">699</span><span class="s2">         if ioargs.encoding and &quot;b&quot; not in ioargs.mode:</span>
<span class="g g-Whitespace">    </span><span class="mi">700</span><span class="s2">             # Encoding</span>
<span class="ne">--&gt; </span><span class="mi">701</span><span class="s2">             handle = open(</span>
<span class="g g-Whitespace">    </span><span class="mi">702</span><span class="s2">                 handle,</span>
<span class="g g-Whitespace">    </span><span class="mi">703</span><span class="s2">                 ioargs.mode,</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] No such file or directory: &#39;FILE_NAME&#39;
</pre></div>
</div>
</div>
</div>
<p><strong>Question:</strong> How many Genes are there in this species and how many different RNAseq libraries were sequenced?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Answer</span>
<span class="c1">## you can do this several ways, the simplest of which is to use the .shape function in pandas to see</span>
<span class="c1">## the demensionality of the data</span>
</pre></div>
</div>
</div>
</div>
<p>You will notice we have three biological replicates in this dataframe (_R1, _R2, _R3). We can assess how similar they are to each other using a principle component analysis. We have a lot of individual data points here (i.e. genes) for doing a PCA, but some are more informative than others. If you look through the data you will see that a lot of genes have no or very little expression. These genes would have TPMs of 0 or &gt;5. Since these data points are not informative, we can mask them before moving on:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filtering the dataframe for genes with low/no expression</span>


<span class="c1">## Filter out non-expressed genes</span>


<span class="c1">## Filter out lowly expressed genes</span>
</pre></div>
</div>
</div>
</div>
<p>TPMs range by several orders of magnitude and a highly expressed gene can have a TMP of &gt; 10,000! If we include values ranging from 1-10,000 in the same data frame, this would skew the PCA. Instead we can log2 transform the data to get numbers that are more reasonable. Note that some genes have 0 expression and log transformation of 0 is undefined. Even though we should have filtered these out, we can add 1 to every value. We can then use the numpy function <code class="docutils literal notranslate"><span class="pre">np.log</span></code> to log transform the values. We will use two loops to do this, as shown in the video.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a loop (s) to log 2 transform the data</span>
</pre></div>
</div>
</div>
</div>
<p>Now our data is in the same order of magnitude and we’re ready to run the PCA. See the transformed values:</p>
<p>If we run the PCA on the data frame as is, we would get one point for each gene. Lets transpose the data so we run the PCA on each sample instead:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the function .transpose() to transpose the dataframe</span>
</pre></div>
</div>
</div>
</div>
<p>Lets run the PCA using the python library scikit-learn. Scikit-learn is a powerful package for machine learning, dimensionality reduction, and predicitve data analysis. This might be a useful for the final project. Today, we will be using a simple fuction in scikit-learn to run a PCA. More details on this code can be found in ‘Notebook 10’ on the class GitHub page (we will not be going over this notebook in class, but it might be worthwhile to go throguh on your own). :ets see how much variation our first two PCs capture. <strong>Note:</strong> This is a complex data frame with many data points that we heavily transformed, so the amount of variation captured by the first two principle components might not be super high.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>
</pre></div>
</div>
</div>
</div>
<p>How much variation is explained by the first two principle components?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the function .explained_variance_ratio_</span>
</pre></div>
</div>
</div>
</div>
<p>Now, we need to add back the sample names into a single data frame and plot the two PCs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add a column to the dataframe containing the sample IDs</span>
</pre></div>
</div>
</div>
</div>
<p>Now that we’ve run the PCA, we need to plot it! There are many ways to make a scatter plot in Python. We wil use the Seaborn library:
<a class="reference external" href="https://seaborn.pydata.org/">https://seaborn.pydata.org/</a> <br>
Seaborn is a data visualization library based on matplotlib, but it has improved functionality (it can make more complex and elegant plots) and it generally makes prettier plots than matplotlib. If you want ot learn more about Seaborn see the Grapevines in a Warming World notebook on the Plants and Python GitHub page: <a class="reference external" href="https://github.com/DanChitwood/PlantsAndPython/blob/master/PlantsAndPython9_STUDENT_Grapevines_in_a_Warming_World.ipynb">https://github.com/DanChitwood/PlantsAndPython/blob/master/PlantsAndPython9_STUDENT_Grapevines_in_a_Warming_World.ipynb</a> <br></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the first two principle components, how do your samples look? </span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="challenge-problem-rnaseq-pipeline">
<h1>Challenge problem: RNAseq pipeline<a class="headerlink" href="#challenge-problem-rnaseq-pipeline" title="Permalink to this headline">¶</a></h1>
<p>So far we have learned how to process RNAseq data using one sample at a time or using a loop in a shell script. We downloaded data from the sequence read archive (SRA), trimmed the reads using <code class="docutils literal notranslate"><span class="pre">fastp</span></code>, aligned the data using <code class="docutils literal notranslate"><span class="pre">Salmon</span></code>, and ran some quality control statistics using a gene expression matrix with a principle component analysis. This gene expression matrix is a starting point for downstream analyses, and there are many analyses we can run to gain some biological insight. We can run differential expression analysis to identify genes that have different expression patterns between tissues, stresses, or developmental timepoints. For instance, we could ask what genes are expressed in response to pathogen attack? We can also do more sophisticated analyses such as a gene network analysis to identify pathways and co-expression modules that are involved in a biological process. We can also do downstream functional analyses to identify what genes that are differentially expressed are doing in the plant using a gene ontology (GO) enrichment test or other similar analyses. <br> <br>
In a typical RNAseq experiment you might have dozens to hundreds of samples, and running them one at a time through half a dozen different programs can be a headache. In this case, it’s easier to run a pipeline with all of the programs strung together where you can supply raw data samples as input and output fully analyzed dataframes ready for the downstream analyses described above. There are several pipelines for RNAseq analyses, and we have created one using our favorite programs to trim, align, quantify, and identify differential expression between samples: <br> <br>
<a class="reference external" href="https://github.com/pardojer23/RNAseqV2">https://github.com/pardojer23/RNAseqV2</a>
<br> <br>
This RNAseq pipeline was created by PhD student Jeremy Pardo, and uses several of the programs we discussed last lesson including <code class="docutils literal notranslate"><span class="pre">fastp</span></code> and <code class="docutils literal notranslate"><span class="pre">salmon</span></code>, and it uses NumPy arrays and pandas to process the data. The RNAseq pipeline will also run differential gene expression analyses using <code class="docutils literal notranslate"><span class="pre">DEseq</span></code>. You simply need to provide it the directory where all of the raw RNAseq files are (2) a sample table a that describes the experimental set up, (3) and the set of transcripts from the species you are working with. The github page provides detailed instructions for how to install and run the pipeline using a conda virtual environment. Because of time, we will not be installing or running the pipeline this summer, but please feel free to install and run it if interested!</p>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "PlantsAndPython/PlantsAndPython",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./Q_10_RNASEQ_ANALYSIS"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="0_Lessons/10.0_RNAseq_analysis.html" title="previous page">10.0 A short introduction to short read analysis</a>
    <a class='right-next' id="next-link" href="../R_11_VARIANT_DISCOVERY/0_Overview.html" title="next page">Welcome / Bienvenid&#64;</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By  <br/>
        
            &copy; Copyright By the authors.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>