{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <p style=\"color:#2C6A7F\">6.2 Biological Data Formats\n",
    "\n",
    "***Author:  Dr. Alejandra Rougon*** \n",
    "\n",
    "<p style=\"color:#9D994B\"><a rel=\"license\" href=\"http://creativecommons.org/licenses/by-nc/4.0/\"><img alt=\"Creative Commons License\" style=\"border-width:0\" src=\"https://i.creativecommons.org/l/by-nc/4.0/88x31.png\" /></a><br />This work is licensed under a <a rel=\"license\" href=\"http://creativecommons.org/licenses/by-nc/4.0/\">Creative Commons Attribution-NonCommercial 4.0 International License</a>.\n",
    "\n",
    "> ### <p style=\"color:#2C6A7F\"> 🔍 **Learning Objectives**\n",
    ">After completing this lesson you will learn:\n",
    ">\n",
    ">* to recognize the most common biological data formats\n",
    ">* DNA and protein sequences [fasta]\n",
    ">* sequencing format [fastq]\n",
    ">* mapping formats [sam, bam]\n",
    ">* annotation formats [bed, gtf, gff, vcf]\n",
    ">* BLAST alignment results in tabular format\n",
    "\n",
    "\n",
    "\n",
    " [![09. Biological Data Formats](https://github.com/alerougon/ObjectStorage/blob/main/PP_CommandLine/MiniaturasVideos/Slide9.png?raw=true)](https://youtube.com/embed/76v6bI5DEXk \"09. Biological Data Formats\")\n",
    "\n",
    "\n",
    "\n",
    "Most biological data is stored in text files. DNA  sequence data is shown in letters `A``G``C``T`. Each character represents the four nucleotides (bases) *adenine, guanine, cytosine and thymine* respectively. Similarly protein data is represented by letters, where one character is assigned to each amino acid. The letters can be either uppercase or lowercase. When you have a mixture of upper and lowercase letters in the same file it usually means the small letters are **masked bases** and they won't be taken into account for some processes by specific programs. There are specific letters to represent ambiguos sequences, in other words, where there are various options of nucleotides or amino acids for that particular position. Also if there is an unknown nucleotide it is represented with the letter `N`. If there is an unknown amino acid residue it is represented by an `X`.\n",
    "\n",
    "| Amino acid                  | Three letter code | One letter code |\n",
    "|-----------------------------|:-------------------:|:-----------------:|\n",
    "| alanine                     | ala               | A               |\n",
    "| arginine                    | arg               | R               |\n",
    "| asparagine                  | asn               | N               |\n",
    "| aspartic acid               | asp               | D               |\n",
    "| asparagine or aspartic acid | asx               | B               |\n",
    "| cysteine                    | cys               | C               |\n",
    "| glutamic acid               | glu               | E               |\n",
    "| glutamine                   | gln               | Q               |\n",
    "| glutamine or glutamic acid  | glx               | Z               |\n",
    "| glycine                     | gly               | G               |\n",
    "| histidine                   | his               | H               |\n",
    "| isoleucine                  | ile               | I               |\n",
    "| leucine                     | leu               | L               |\n",
    "| lysine                      | lys               | K               |\n",
    "| methionine                  | met               | M               |\n",
    "| phenylalanine               | phe               | F               |\n",
    "| proline                     | pro               | P               |\n",
    "| serine                      | ser               | S               |\n",
    "| threonine                   | thr               | T               |\n",
    "| tryptophan                  | trp               | W               |\n",
    "| tyrosine                    | tyr               | Y               |\n",
    "| valine                      | val               | V               |\n",
    "\n",
    "\n",
    "## <p style=\"color:#2C6A7F\"> FASTA\n",
    "\n",
    "\n",
    "The most common format to store DNA and protein sequences is the `fasta` format, which was originated from a program that bears the same name. The `.fasta` file contains a `>` sign followed by the sequence identifier and in the next line the sequence in nucleotides or amino acids (represented with a single letter). The sequence could be in a single line\n",
    "\n",
    "```\n",
    ">sequence_1\n",
    "ATTTATGGCCCGCGATGTACGCCATCCAGACTTAACAGTGGGACATGGCACACAGTTGAGAGGGCACAAAATATTTAGACAGATAC\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or in multiple lines, usually of 60 characters each\n",
    "\n",
    "```\n",
    ">sequence_2\n",
    "AAGCTTCGGCATAGGACAAGATGGAGAGGGACGCATTGTGGAAGGCGCGCTGGGTGTTTTGGCAACCAGC\n",
    "GCAGTTGCCTGTTTGTGGTTAACATAGCTGTTACGAGAAGACGAGCGGCAAAGCTGTCATGCGTCTTGCT\n",
    "GTGCGAACGCAGTCGTCGGGCGAATTGTCCGTACGCGGGTGGTGGGGGGTGCAAATAGCCTGCGCTTGCA\n",
    "TTTCAGCCTGGGCATTCGAGTGGAATATTATTGTGTTACACTGTACAAAAAAAAAAAGCTT\n",
    "```\n",
    "Records in a fasta file can also have descriptions. They will come after the id, usually separated by a space.\n",
    "\n",
    "```\n",
    ">QBF53359.1 RxLR [Phytophthora capsici]\n",
    "MSKVFLLLVLSVFALVSCDALSAPVGSKLSLSKTDELNAQPIDAKRMLRAQEEPTNAADEERGMTELANK\n",
    "FKAWAAAIKTWVTNSKLVQSMNNKLASLTQKGRVGQIEKLLKQDNVNVNVLYQNKVKPDELFLALKLDPK\n",
    "LKLIADAPAAWANNPGLSMFYQYATYYAKMTTKA\n",
    "```\n",
    "\n",
    "You can also have multiple sequences in a fasta file\n",
    "\n",
    "```\n",
    ">QBF53359.1 RxLR [Phytophthora capsici]\n",
    "MSKVFLLLVLSVFALVSCDALSAPVGSKLSLSKTDELNAQPIDAKRMLRAQEEPTNAADEERGMTELANK\n",
    "FKAWAAAIKTWVTNSKLVQSMNNKLASLTQKGRVGQIEKLLKQDNVNVNVLYQNKVKPDELFLALKLDPK\n",
    "LKLIADAPAAWANNPGLSMFYQYATYYAKMTTKA\n",
    "\n",
    ">sp|D0N607.1|CRE4_PHYIT RecName: Full=RxLR effector protein CRE4; AltName: Full=Core RXLR effector 4; Flags: Precursor\n",
    "MLRSFLLIVATVSLFGQCKPLPLATSPVSDAVRAPHRSTHETRFVRTNDEERGATMTLAGVLRDKAQTKQ\n",
    "LLTSWLNSGKSVPSVSNKLGLKRMSLEQAIHHENWKALTTFQRMKSKKAKAYAKYGTGYQTEAKTKENLL\n",
    "QWVMRGDSPKEVSSTLGLLGLSRRKIIDHQNYEAFRTFLKYRKQWAEMQGNGFTKLTT\n",
    "\n",
    ">sp|D0MRS3.1|RD21_PHYIT RecName: Full=RxLR effector protein PexRD21; AltName: Full=Core RXLR effector 1; Flags: Precursor\n",
    "MRLSYILVVVIAVTLQACVCATPVIKEANQAMLANGPLPSIVNTEGGRLLRGVKKRTAEREVQEERMSGA\n",
    "KLSEKGKQFLKWFFRGSDTRVKGRSWR\n",
    "```\n",
    "The most common extensions used for these files are `.fasta`, `.fa`, `.fna`[for nucleotides], `.faa` [for amino acid residues], `.fsa`, or `.fas`.\n",
    "\n",
    "## <p style=\"color:#2C6A7F\"> FASTQ\n",
    "\n",
    "Next generation sequencing data can come in various formats, however, the most common format is **fastq** and the other formats are usually converted to fastq for their analysis. \n",
    "\n",
    "### Characteristics of the `fastq` format\n",
    "\n",
    "* Each record always contains **four** lines\n",
    "* The **first line** starts with a `@` sign followed by the **sequence identifier** \n",
    "* The contents of the first line vary depending on the sequencing machine, version and the conversion software. However, it usually contains information about the sequencing run and cluster location.\n",
    "* The **second line** contains the **sequence** or **base calls (A,C,T,G)** or N if the base couldn't be identified.\n",
    "* The **third line** contains a `+` sign and in some cases it repeats the sequence identifier of line one.\n",
    "* The **fourth line** contains the **quality scores** of each base in ASCII format \n",
    "\n",
    "```\n",
    "@<identifier and expected information>\n",
    "<sequence>\n",
    "+<identifier and other information OR empty string>\n",
    "<quality>\n",
    "```\n",
    "\n",
    "specific information of the data contained in the description of the first line can be found [here](https://help.basespace.illumina.com/articles/descriptive/fastq-files/)\n",
    "\n",
    "### Example of a single fastq record:\n",
    "\n",
    "```\n",
    "@SIM:1:FCX:1:15:6329:1045 1:N:0:2\n",
    "TCGCACTCAACGCCCTGCATATGACAAGACAGAATC\n",
    "+\n",
    "<>;##=><9=AAAAAAAAAA9#:<#<;<<<????#=\n",
    "\n",
    "```\n",
    "Each record represents a sequencing read\n",
    "\n",
    "### Paired libraries\n",
    "In paired-end or mate-paired libraries we obtain two sequences that are associated to the same record [they have the same identifier]. In fact, they come from both extremes of a single DNA fragment. Both pairs can come in **separated files** or in the same file, **interleaved**. The most common format is in separate files. In either case the pairs (that have the same identifier) have to be differentiated. \n",
    "\n",
    "Interleaved files have first the forward `F` pair, followed by the reverse pair `R`. And when they are in separate files they are indicated in the first line as `/1` and `/2` or in newer versions `1:Y:18:ATCACG` and `2:Y:18:ATCACG`\n",
    "\n",
    "**Example of file with the first pair of each record** `1:N:0:1`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "@M02586:164:000000000-AYBV7:1:2119:13975:24958 1:N:0:1\n",
    "ACGTTGTGCCAGCCGGGTGAGAAAGTAGCTGTCCGTTTCGCTGGTCTGGTCAATATGCGGGATCTTAACGCTATCAAGTTCCTGCGGGATTACTGGCAGATCTAATTTATTCCGGCTTGAAATGATTCTGACAATCGCGCCGAGCGTGTAATCATCGTAAGACTGG\n",
    "+\n",
    "CCCCCGGGGGGGFGGGGGGFFEGGGGGGGFCFEEGFGFFFGGGGGGGGGGGGGGFGGFG@:CFGGGGGGGGGGGGGGFFGGFFGGDGGGGGGGGFGFGGFGGGGGGGGGGD<FGGGGGGGGGGGGGGGFFFFGFF8CGGDFFEGGGGGGFGGDFGGG<DCFGGCFG\n",
    "@M02586:164:000000000-AYBV7:1:2119:13272:24958 1:N:0:1\n",
    "CACTCACTAGCGACCCACTTGGAGGTCGCCAAAAGCAATATTATCCATATCTCGGAGGATTTGGACCAGATCCAGGTCGTCATCGAATTGCCTGGAGAACGGCACGAACTTGGTGCTACGACGAAGTGGAAACGCACAGCAACTTCTGGAGGGAGTGTTGGATCGA\n",
    "+\n",
    "CCCCCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGDFFGGGGGGGGGGGGGGGGGGGGGGGGCFGGGGGGGGGGDGGGGGGGGGGGGFFG:@GGFGGGGGGG5>,\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "**Example of file with the second pair of each record** `2:N:0:1`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "@M02586:164:000000000-AYBV7:1:2119:13975:24958 2:N:0:1\n",
    "TNNNNNNNNNCAGTTACCCGTGGGGGCTCGCATCGCACCCGCATTCACGCTGACGATAAAAAATAAGGTGCTGGAAGAAAATATCTCTGCTCGGATCATCAGTTTATCTGTCACGGATAACAGCGGTTTTACCGCAGACACCTTGAATTTAACCTTCGATGACAGC\n",
    "+\n",
    "8#########==CFCEDCFEGGGGGGCCGFGEGEGGCGGGGGGGD9@FFCFGFDGEGCFFCCCFGGEFGG8FFGGGFGFFFCCFA<FFGFGGGG:FEFGGFGGGFGGFGGGGA?CCFGGFFGGEG7CF@FGEF8+=CFGFEFF:=CFGGGAFGFFG6C?(*-1*)/\n",
    "@M02586:164:000000000-AYBV7:1:2119:13272:24958 2:N:0:1\n",
    "ANNNNNNNNNCTGGAAATTGTAACCCATAGGCGGTCCATGTGCGTTTCGTTCTGAGATCGTTTTTTCGTACATCGTTGACAAGCCCAAGTTCTCACCACAAGATTCGTTAGGCTCGTATCGTAATGTGGAAAGGTAGACAGGCGATTCGAGCGAGATCAAGTCGAC\n",
    "+\n",
    "8#########=:CFGGFFGGGGGGGGG8FFGGGGFGGGGGG9FEGGGGGGGGFGGGGGGGF@FGGGGGGGGGGGGGGGGGFFGGEFGGGGGFGGCGGGGGEGFFGFFGFF7FFE<FCGGGGFEF@FGFF9EF<FFG,?;<F+DEGGCCGGGGECE*.4)72*)-5(\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## <p style=\"color:#2C6A7F\"> BAM & SAM\n",
    "\n",
    "When sequencing a genome, one has to fractionate the DNA so pieces of the genome get sequenced. The result is a file with millions of sequences that have to be assembled back to form a genome. The fragments sizes depend on the technology that we have used. Second generation technologies produce small fragments (~35-400bp). Currently the most common sizes are 50-300bp from Illumina reads. Third generation sequencing technologies (Long Read Sequencing) provide much longer fragments (500bp-2.3Mb although they are usually 10-30kb). The task of reassembling a genome is usually not completely fulfilled. What we get are bigger fragments than those initially sequenced called contigs or scaffolds. \n",
    "\n",
    "If we already have a reference genome we can align [map] those sequences [reads] to the reference genome. The most common formats to show the positions at which each read has aligned in relation to the reference sequence are the **sam** format and its compressed binary version, the **bam** format. The bam format cannot be opened directly with a text editor as is in binary.\n",
    "\n",
    "![](https://github.com/alerougon/ObjectStorage/blob/main/PP_CommandLine/24.MapReads.png?raw=true)\n",
    "\n",
    "The sam format contains two sections. One is the **header** which is identified with an `@` and two Uppercase letters and then the **alignment information section** that contains the positions at which each read has aligned. The header position may be removed for analysis. \n",
    "The alignment information is divided into 11 columns. Additionally some optional **tags** can be added for more alignment specifications.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "@HD     VN:1.0  SO:unsorted\n",
    "@SQ     SN:12   LN:50697278\n",
    "@PG     ID:bowtie2 PN:bowtie2  VN:2.2.5  CL:\"bowtie2 -x ZV9 -S 2cells.sam -1 1.fq -2 2.fq\"\n",
    "\n",
    "ERR022484.110   137     chr12      4627377 42      76M     =       4627377 0       TTGTTCTCCACCAAGCCGCCCAGTTT    EEEEEEEEEEEEEE@EBBEEE;C AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:76 YT:Z:UP\n",
    "ERR022484.110   69      chr12      4627377 0       *       =       4627377 0       CNGAAGCAAAGTGTGTGCGCGAAATG    !AA=CCCDBHAG=ADHADD>D>D    YT:Z:UP\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "The main columns contain the following data. \n",
    "\n",
    "| Col | Field | Brief description                     |\n",
    "|-----|-------|---------------------------------------|\n",
    "| 1   | QNAME | Query template NAME                   |\n",
    "| 2   | FLAG  | bitwise FLAG                          |\n",
    "| 3   | RNAME | Reference sequence NAME               |\n",
    "| 4   | POS   | leftmost mapping POSition             |\n",
    "| 5   | MAPQ  | MAPping Quality                       |\n",
    "| 6   | CIGAR | CIGAR string                          |\n",
    "| 7   | RNEXT | Reference name of the mate/next read  |\n",
    "| 8   | PNEXT | Position of the mate/next read        |\n",
    "| 9   | TLEN  | observed Template LENgth              |\n",
    "| 10  | SEQ   | segment SEQuence                      |\n",
    "| 11  | QUAL  | ASCII of Phred-scaled base QUALity+33 |\n",
    "\n",
    "We won't enter into detail of what the codes mean in the sam format. For further details you can check the [format specifications](http://samtools.github.io/hts-specs/SAMv1.pdf) and the [optional tags descriptions](https://samtools.github.io/hts-specs/SAMtags.pdf).\n",
    "\n",
    "## <p style=\"color:#2C6A7F\"> Annotation formats\n",
    "\n",
    "Genome **annotation** is the process of identifying the **locations of genes** and all of the coding regions in a genome [structural annotation] and determining what those genes do [functional annotation]. There are several formats for representing the locations of genes in a genome. Annotation formats contain the name of the gene, the positions at which each gene starts and ends in reference to the genomic sequence, and the positions of the different parts of the gene **[features]** that have been annotated [UTRs, Exons, Promoters, etc]. Some genes can have various forms **[transcripts]** due to **alternative splicing**. The genomic sequence can be in scaffolds or chromosomes. Here you can see some examples of different annotation formats. All of them are basically tables, contain columns that are usually separated by a `tab` and rows. Tables are in **tabular** format. Annotation formats are flexible. They vary since programs may require slight modifications. These are just general examples.\n",
    "\n",
    "\n",
    "**BED** [Browser Extensible Data]\n",
    "Basic format [columns 1-3 are required]. Genome start at position 0 [0-based]. \n",
    "\n",
    "```\n",
    "#chr\tstart end\n",
    "chr12\t\t10134\t\t10600\t\t\n",
    "chr12\t\t10977\t\t11008\n",
    "chr12\t\t13409\t\t14312\n",
    "```\n",
    "**BED** Extended format - groups [Each line represents a transcript].\n",
    "\n",
    "```\n",
    "#chr start end name score strand block_count(number of exons) block_sizes(each exon size)\n",
    "chr22 1000 5000 cloneA 960 + 1000 5000 0 2 567,488, 0,3512\n",
    "chr22 2000 6000 cloneB 900 - 2000 6000 0 2 433,399, 0,3601\n",
    "```\n",
    "**BED** Extended format - rgb  [indicates color of thick lines of transcripts in a genome browser].\n",
    "\n",
    "```\n",
    "#chr\tstart end\tname\tscore\tstrand\tthick_start\tthick_end rb\n",
    "chr12\t10134\t10600\tExon1 100\t+\t10180\t10600\t255,0,0\n",
    "chr12\t10977\t11008\tExon2 100 + 10977\t11000 255,0,0\n",
    "chr12\t13409\t14312\tExon3 100 + 13409\t14300 255,0,0\n",
    "chr12\t18114\t18423\tExon4 100 - 18150\t18423 0,0,255\n",
    "```\n",
    "\n",
    "**GTF** [Gene Transfer Format]\n",
    "Genome starts at position 1 [1-base]. Each line represents a feature. The whole transcript is represented in several lines containing all its features.\n",
    "\n",
    "```\n",
    "#chr\tprogram feature start end strand  frame  gene_id; transcript_id\n",
    "chr12\tGF\tExon\t10135\t10600\t100\t+\t.\tgene_id \"genA\"; transcript_id\t\"geneA.1\";\n",
    "chr12\tGF\tExon\t10978\t11008\t100\t+\t.\tgene_id\t\"genA\";\ttranscript_id\t\"geneA.1\";\n",
    "chr12\tGF\tExon\t13410\t14312\t100\t-\t.\tgene_id\t\"genB\";\ttranscript_id\t\"geneB.1\";\n",
    "chr12\tGF\tExon\t18114\t18423\t100\t-\t.\tgene_id\t\"genB\";\ttranscript_id\t\"geneB.1\";\n",
    "```\n",
    "\n",
    "**GFF** [General Feature Format] \n",
    "Genome starts at position 1 [1-base]. Each line represents a feature. \n",
    "\n",
    "```\n",
    "##chr\tprogram feature start end score\tstrand  frame  attributes\n",
    "chr3 GF mRNA            1300  9000  .  +  .  ID=mrna0001;Name=sonichedgehog\n",
    "chr3 GF exon            1300  1500  .  +  .  ID=exon00001;Parent=mrna0001\n",
    "chr3 GF exon            1050  1500  .  +  .  ID=exon00002;Parent=mrna0001\n",
    "chr3 GF exon            3000  3902  .  +  .  ID=exon00003;Parent=mrna0001\n",
    "```\n",
    "\n",
    "## <p style=\"color:#2C6A7F\"> Sequence variation\n",
    "\n",
    "Each individual's genome sequence is unique. The differences in the DNA sequences of individuals are ultimately responsible for differences in observable traits, such as  eye color or height, as well as for the hidden ones. For instance, that differences may determine how susceptible or resistant a plant is to a disease, or how virulent a specific pathogen is. \n",
    "\n",
    "In humans, there are 0.1% differences between the genomes of any two individuals.  That means, that out of a three billion base sequence, there is roughly three million differences between any two individuals. This variation can be due to substitutions, insertions or deletions. The **VCF** [Variant Call Format], and its Binary format **BCF**  are  used for giving information about sequence variation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "##fileformat=VCFv4.1\n",
    "##ApplyRecalibration=\"analysis_type=ApplyRecalibration input_file=[] read_buffer_size=null phone_home=NO_ET gatk_key=/\n",
    "##CombineVariants=\"analysis_type=CombineVariants input_file=[] read_buffer_size=null phone_home=STANDARD \n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
    "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">\n",
    "##contig=<ID=chr1,length=101,assembly=Ddip,md5=bd01f7e11515bb6beda8f7257902aa67>\n",
    "##contig=<ID=chr2,length=101,assembly=Ddip,md5=31c33e2155b3de5e2554b693c475b310>\n",
    "##contig=<ID=chr3,length=101,assembly=Ddip,md5=631593c6dd2048ae88dcce2bd505d295>\n",
    "##contig=<ID=chr4,length=101,assembly=Ddip,md5=c60cb92f1ee5b78053c92bdbfa19abf1>\n",
    "##source= Ddip haplotype map vcf for testing\n",
    "#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  Ddip1\n",
    "chr1    75      Dd1     C       A       .       .       .       GT      0/0\n",
    "chr3    25      Dd9     C       T       .       .       .       GT      0/1\n",
    "chr3    75      Dd6     A       T       .       .       .       GT      1/1\n",
    "chr3    100     Dd8     A       T       .       .       .       GT:PS   0/0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## <p style=\"color:#2C6A7F\"> Sequence alignments\n",
    "After having a genome structurally annotated we may want to know the functions of each gene. Specially if we have found variation that may represent a significant change in the fenotype of an individual. For example, a variation that may represent the loss of function of a disease resistant gene. The process of finding the function of the genes is called **functional annotation** and it involves the aligning of the genes or proteins to a database to try to find homologies to already annotated proteins. The results of those alignments can be obtained in a **tabular** format; as we saw earlier, this means, a table, with rows and columns. \n",
    "\n",
    "Here is an example of an alignment obtained after running a program called **BLAST**\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "# BLASTP 2.5.0+\n",
    "# Query: gi|49146530|ref|YP_026090.1| NADH dehydrogenase subunit 5 (mitochondrion) [Steinernema carpocapsae]\n",
    "# Database: PlusMitoDB\n",
    "# Fields: query acc., subject acc., % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
    "# 500 hits found\n",
    "gi|49146530|ref|YP_026090.1|\tgi|49146530|ref|YP_026090.1|\t100.000\t527\t0\t0\t1\t527\t1\t527\t0.0\t981\n",
    "gi|49146530|ref|YP_026090.1|\tgi|910356121|ref|YP_009161998.1|\t71.619\t525\t149\t0\t1\t525\t1\t525\t0.0\t709\n",
    "gi|49146530|ref|YP_026090.1|\tgi|910356106|ref|YP_009161984.1|\t70.476\t525\t155\t0\t1\t525\t1\t525\t0.0\t709\n",
    "gi|49146530|ref|YP_026090.1|\tgi|910356145|ref|YP_009162020.1|\t70.857\t525\t153\t0\t1\t525\t1\t525\t0.0\t708\n",
    "gi|49146530|ref|YP_026090.1|\tgi|116510842|ref|YP_817460.1|\t70.857\t525\t153\t0\t1\t525\t1\t525\t0.0\t701\n",
    "gi|49146530|ref|YP_026090.1|\tgi|910356132|ref|YP_009162008.1|\t72.800\t500\t136\t0\t26\t525\t27\t526\t0.0\t686\n",
    "gi|49146530|ref|YP_026090.1|\tgi|5834894|ref|NP_006964.1|ND5_10021\t69.714\t525\t159\t0\t1\t525\t1\t525\t0.0\t674\n",
    "gi|49146530|ref|YP_026090.1|\tgi|188011122|ref|YP_001905895.1|\t70.611\t524\t154\t0\t1\t524\t1\t524\t0.0\t669\n",
    "gi|49146530|ref|YP_026090.1|\tgi|620695076|ref|YP_009027244.1|\t69.466\t524\t160\t0\t1\t524\t1\t524\t0.0\t669\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------\n",
    "🔑 **In this section you have learned**\n",
    "\n",
    "* to recognize the most common biological data formats\n",
    "* DNA and protein sequences [fasta]\n",
    "* Sequencing results [fastq]\n",
    "* Mapping formats [sam, bam]\n",
    "* Annotation formats [bed, gtf, gff, vcf]\n",
    "* BLAST alignment results in tabular format\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}