
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>10.0 A short introduction to short read analysis &#8212; Plants &amp; Python</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet" />
  <link href="../../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe,.cell"
        const thebe_selector_input = "pre,.cell_input div.highlight"
        const thebe_selector_output = ".output,.cell_output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Practice: Introduction to gene expression (RNAseq) analysis" href="../1_Practice.html" />
    <link rel="prev" title="Welcome / Bienvenid@" href="../0_Overview.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      <img src="../../_static/plants_python_logo.jpg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Plants & Python</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../00_Opening_page.html">
   Welcome to / Bienvenid@ a Plants &amp; Python
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Preface
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../00_PREFACE/01_Overview.html">
   Welcome : Bienvenid@
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/01_Welcome_Bienvenido/00_Welcome.html">
     Welcome to Plants &amp; Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/01_Welcome_Bienvenido/01_Bienvenido.html">
     Bienvenid@ a Plants &amp; Python
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../00_PREFACE/Introduction_Illustration.html">
   Introduction Illustration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../00_PREFACE/Ilustracion_de_introduccion.html">
   Ilustración de introducción
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../00_PREFACE/02_Overview.html">
   Getting Started : Empezando
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/00_Installing_Anaconda.html">
     1. Installing Anaconda and using Jupyter notebooks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/01_Instalar_Anaconda.html">
     1. Instalar Anaconda y usar Jupyter notebooks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/02_Coding_cells.html">
     2. Coding cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/03_Celdas_de_codigo.html">
     2. Celdas de código
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/04_Markdown_cells.html">
     3. Markdown cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/02_Getting_started_Empezando/05_Celdas_de_markdown.html">
     3. Celdas de markdown
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../00_PREFACE/03_Overview.html">
   Using : Usando Jupyter Book
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/03_Using_Usando_Jupyter_Book/How_to_use_this_Jupyter_Book.html">
     How to use this Jupyter Book
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../00_PREFACE/03_Using_Usando_Jupyter_Book/Como_usar_este_Jupyter_Book.html">
     Cómo usar este Jupyter Book
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  1. Variables, lists, &amp; indexing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/Cover_image_1.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.0_Variables.html">
     1.0 Variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.1_Functions.html">
     1.1 Functions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.2_Indexing.html">
     1.2 Indexing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.3_Lists.html">
     1.3 Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.4_Indexing_List_Elements.html">
     1.4 Indexing List Elements
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.5_Modifying_Lists.html">
     1.5 Modifying Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.6_List_of_Lists.html">
     1.6 List of Lists
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.7_Slicing.html">
     1.7 Slicing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/0_Lessons/1.8_Data_Types.html">
     1.8 Data Types
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../A_1_VARIABLES_LISTS_AND_INDEXING/2_Plants_And_Climate.html">
   Plants &amp; Climate
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  1. Variables, listas, &amp; INDEXACIÓN
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/Ilustracion_de_portada_1.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.0_Variables.html">
     1.0 Variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.1_Funciones.html">
     1.1 Funciones
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.2_Indexacion.html">
     1.2 Indexación
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.3_Listas.html">
     1.3 Listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.4_Indexando_los_elementos_de_una_lista.html">
     1.4 Indexando los elementos de una lista
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.5_Modificando_listas.html">
     1.5 Modificando listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.6_Lista_de_listas.html">
     1.6 Lista de listas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.7_Rebanar.html">
     1.7 Rebanar
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/0_Lecciones/1.8_Tipos_de_datos.html">
     1.8 Tipos de datos
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../B_1_VARIABLES_LISTAS_Y_INDEXACION/2_Las_plantas_y_el_clima.html">
   Las plantas y el clima
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  2. Visualizing Data with Matplotlib
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/Cover_image_2.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.0_Modules.html">
     2.0 Modules
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.1_Plots.html">
     2.1 Plots
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.2_Shape_Size_and_Color.html">
     2.2 Shape, Size, and Color
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.3_Legends.html">
     2.3 Legends
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/0_Lessons/2.4_Subplots.html">
     2.4 Subplots
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../C_2_VISUALIZING_DATA_WITH_MATPLOTLIB/2_Leaf_And_Sunflower_Shapes.html">
   Leaf &amp; Sunflower Shapes
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  2. VISUALIZACIÓN DE DATOS CON MATPLOTLIB
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/Ilustracion_de_portada_2.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.0_Modulos.html">
     2.0 Módulos
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.1_Graficos.html">
     2.1 Gráficos
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.2_Forma_tamano_y_color.html">
     2.2 Forma, tamaño y color
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.3_Leyendas.html">
     2.3 Leyendas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/0_Lecciones/2.4_Subgraficos.html">
     2.4 Subgráficos
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../D_2_VISUALIZACION_DE_DATOS_CON_MATPLOTLIB/2_Las_formas_de_las_hojas_y_girasoles.html">
   Las formas de las hojas y girasoles
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  3. Loops &amp; the Golden Angle
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/Cover_image_3.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.0_Range_Function.html">
     3.0 Range Function
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.1_For_Loops.html">
     3.1 For Loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.2_Counters.html">
     3.2 Counters
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.3_Using_Lists_and_Loops.html">
     3.3 Using Lists and Loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.4_Calculating_Fibonacci.html">
     3.4 Calculating Fibonacci
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/0_Lessons/3.5_While_Loops.html">
     3.5 While Loops
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../E_3_LOOPS_AND_THE_GOLDEN_ANGLE/2_Modeling_Sunflowers.html">
   Modeling Sunflowers
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  3. LOOPS &amp; EL ÁNGULO ÁUREO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/Ilustracion_de_portada_3.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.0_La_funcion_range.html">
     3.0 La función range
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.1_Loops_for.html">
     3.1 Loops for
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.2_Contadores.html">
     3.2 Contadores
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.3_Usando_listas_y_loops.html">
     3.3 Usando listas y loops
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.4_Calcular_Fibonacci.html">
     3.4 Calcular Fibonacci
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/0_Lecciones/3.5_Loops_while.html">
     3.5 Loops while
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../F_3_LOOPS_Y_EL_ANGULO_DORADO/2_Como_construir_un_girasol.html">
   Cómo construir un girasol
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  4. Data Analysis with Pandas
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/Cover_image_4.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.0_Reading_in_Data.html">
     4.0 Reading in Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.1_Head_and_Tail.html">
     4.1 Head and Tail
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.2_Selecting_Columns.html">
     4.2 Selecting Columns
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.3_Descriptive_Statistics.html">
     4.3 Descriptive Statistics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.4_Masks.html">
     4.4 Masks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.5_Using_iloc.html">
     4.5 Using iloc
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/0_Lessons/4.6_Seaborn.html">
     4.6 Seaborn
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/1_Practice.html">
   Practice
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../G_4_DATA_ANALYSIS_WITH_PANDAS/2_Grapes_in_a_Warming_World.html">
   Grapes in a Warming World
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  4. Análisis de datos con pandas
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/Ilustracion_de_portada_4.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.0_Leyendo_en_data.html">
     4.0 Leyendo en data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.1_Head_y_tail.html">
     4.1 Head y tail
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.2_Seleccionando_columnas.html">
     4.2 Seleccionando columnas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.3_Estadisticas_descriptivas.html">
     4.3 Estadísticas descriptivas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.4_Enmascaramiento.html">
     4.4 Enmascaramiento
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.5_Usando_iloc.html">
     4.5 Usando iloc
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/0_Lecciones/4.6_Seaborn.html">
     4.6 Seaborn
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/1_Practica.html">
   Práctica
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../H_4_ANALISIS_DE_DATOS_CON_PANDAS/2_Vides_en_un_mundo_que_se_calienta.html">
   Vides en un mundo que se calienta
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  5. Unix Command Line
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/Cover_image_5.html">
   Cover Illustration
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.0_Introduction_to_Command_Line.html">
     <p style="color:#2C6A7F">
      5.0 Introduction to Command Line
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.1_TheVirtual_Terminal.html">
     <p style="color:#2C6A7F">
      5.1 Virtual Terminal
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.2_Exploring_Folders.html">
     <p style="color:#2C6A7F">
      5.2 Exploring Folders
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.3_Working_with_Files.html">
     <p style="color:#2C6A7F">
      5.3 Working with Files
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.4_Options_and_Arguments.html">
     <p style="color:#2C6A7F">
      5.4 Options and Arguments
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/0_Lessons/5.5_Copy_Move_and_Remove_Files.html">
     <p style="color:#2C6A7F">
      5.5 Copy, Move &amp; Remove Files
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../I_5_UNIX_COMMAND_LINE/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  5. LINEA DE COMANDOS DE UNIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/Ilustracion_de_portada_5.html">
   Ilustración de portada
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.0_IntroduccionLineaDeComandos.html">
     <p style="color:#2C6A7F">
      5.0 Introducción a la línea de comandos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.1_LaTerminal_Virtual.html">
     <p style="color:#2C6A7F">
      5.1 Terminal Virtual
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.2_ExplorandoDirectorios.html">
     <p style="color:#2C6A7F">
      5.2 Explorando directorios
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.3_Trabajando_con_archivos.html">
     <p style="color:#2C6A7F">
      5.3 Trabajando con archivos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.4_Opciones_y_Argumentos.html">
     <p style="color:#2C6A7F">
      5.4 Opciones y argumentos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/5.5_Copiar_Mover_Eliminar_Archivos.html">
     <p style="color:#2C6A7F">
      5.5 Copiar, mover y eliminar archivos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../J_5_LINEA_DE_COMANDOS_DE_UNIX/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  6. Unix Commands For Data Mining
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
  <label for="toctree-checkbox-14">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.0_Edit_Sort_Cut_grep.html">
     <p style="color:#2C6A7F">
      6.0 EDIT, SORT, CUT &amp; GREP
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.1_Word_Count_Pipelines_Permissions.html">
     <p style="color:#2C6A7F">
      6.1 Word Count, Pipelines &amp; Permissions
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../K_6_UNIX_COMMANDS_FOR_DATA_MINING/0_Lessons/6.2_Biological_Data_Formats.html">
     <p style="color:#2C6A7F">
      6.2 Biological Data Formats
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../K_6_UNIX_COMMANDS_FOR_DATA_MINING/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  6. MINERIA DE DATOS EN LA LINEA DE COMANDOS DE UNIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
  <label for="toctree-checkbox-15">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.0_Editar_Ordentar_Cortar_grep.html">
     <p style="color:#2C6A7F">
      6.0 Editar, ordenar, cortar y grep
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.1_Contar_Palabras_Tuberias_Permisos.html">
     <p style="color:#2C6A7F">
      6.1 Contar palabras, tuberías &amp; permisos
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/0_Lecciones/6.2_Formatos_Biologicos.html">
     <p style="color:#2C6A7F">
      6.2 Formatos de Datos Biológicos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../L_6_MINERIA_DE_DATOS_EN_LA_LINEA_DE_COMANDOS_DE_UNIX/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  7. SCRIPTING WITH BASH
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../M_7_SCRIPTING_WITH_BASH/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
  <label for="toctree-checkbox-16">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.0_WorkingRemotely.html">
     <p style="color:#2C6A7F">
      7.0 Working Remotely
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.1_Scripting_with_bash.html">
     <p style="color:#2C6A7F">
      7.1 Scripting with Bash
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../M_7_SCRIPTING_WITH_BASH/0_Lessons/7.2_Summary_of_Commands.html">
     <p style="color:#2C6A7F">
      7.2 Summary of Commands
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../M_7_SCRIPTING_WITH_BASH/1_Practice.html">
   Practice
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  7. PROGRAMANDO CON BASH
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../N_7_PROGRAMANDO_CON_BASH/0_Overview.html">
   Lecciones
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
  <label for="toctree-checkbox-17">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.0_Trabajando_de_manera_remota.html">
     <p style="color:#2C6A7F">
      7.0 Trabajando de manera remota
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.1_Programando_con_Bash.html">
     <p style="color:#2C6A7F">
      7.1 Programando con Bash
     </p>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../N_7_PROGRAMANDO_CON_BASH/0_Lecciones/7.2_Resumen_de_comandos.html">
     <p style="color:#2C6A7F">
      7.2 Resumen de comandos
     </p>
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../N_7_PROGRAMANDO_CON_BASH/1_Practica.html">
   Práctica
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  8. GENOMICS AND BIOPYTHON
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../O_8_GENOMICS_AND_BIOPYTHON/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
  <label for="toctree-checkbox-18">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../O_8_GENOMICS_AND_BIOPYTHON/0_Lessons/8.0_Intro_to_Genomics_and_Biopython.html">
     8.0 Introduction to plant genomics and working with sequences in Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../O_8_GENOMICS_AND_BIOPYTHON/0_Lessons/8.1_HPCC.html">
     Lesson 8.6 Using the high performance computing cluster (HPCC) at MSU
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../O_8_GENOMICS_AND_BIOPYTHON/1_Practice.html">
   Practice: How to Build a Genome
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  9. COMPARATIVE GENOMICS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../P_9_COMPARATIVE_GENOMICS/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
  <label for="toctree-checkbox-19">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../P_9_COMPARATIVE_GENOMICS/0_Lessons/9.0_Comp_Genomics.html">
     9.0 Genome visualization and comparative genomics
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../P_9_COMPARATIVE_GENOMICS/1_Practice.html">
   Practice: Genome visualization and comparative genomics
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  10. RNASEQ ANALYSIS
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../0_Overview.html">
   Lessons
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
  <label for="toctree-checkbox-20">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     10.0 A short introduction to short read analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../1_Practice.html">
   Practice: Introduction to gene expression (RNAseq) analysis
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  11. VARIANT DISCOVERY
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../R_11_VARIANT_DISCOVERY/0_Overview.html">
   Lessons
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
  <label for="toctree-checkbox-21">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../R_11_VARIANT_DISCOVERY/0_Lessons/11.0_Variant_Discovery.html">
     11.0 Variant discovery
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../R_11_VARIANT_DISCOVERY/1_Practice.html">
   Practice: Variant discovery, genetic diversity, and population genetics
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/Q_10_RNASEQ_ANALYSIS/0_Lessons/10.0_RNAseq_analysis.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/PlantsAndPython/PlantsAndPython"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/PlantsAndPython/PlantsAndPython/issues/new?title=Issue%20on%20page%20%2FQ_10_RNASEQ_ANALYSIS/0_Lessons/10.0_RNAseq_analysis.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/PlantsAndPython/PlantsAndPython/master?urlpath=tree/Q_10_RNASEQ_ANALYSIS/0_Lessons/10.0_RNAseq_analysis.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/PlantsAndPython/PlantsAndPython/blob/master/Q_10_RNASEQ_ANALYSIS/0_Lessons/10.0_RNAseq_analysis.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        <button type="button" class="btn btn-secondary topbarbtn"
            onclick="initThebeSBT()" title="Launch Thebe" data-toggle="tooltip" data-placement="left"><i
                class="fas fa-play"></i><span style="margin-left: .4em;">Live Code</span></button>
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   10.0 A short introduction to short read analysis
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plant-biology-learning-objectives">
     Plant Biology Learning Objectives:
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#python-learning-objectives">
     Python Learning Objectives:
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction-to-gene-expression-rnaseq-analysis">
   10.1 Introduction to gene expression (RNAseq) analysis
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#downloading-data-from-the-ncbi-sequence-read-archive-sra">
   Downloading data from the NCBI Sequence Read Archive (SRA)
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#challenge-problem-download-sra-data-onto-hpcc">
   Challenge problem: Download SRA data onto HPCC:
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dimensionality-reduction">
   10.2 Dimensionality reduction
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <p><img alt="plants_python_logo.jpg" src="../../_images/plants_python_logo.jpg" /></p>
<div class="section" id="a-short-introduction-to-short-read-analysis">
<h1>10.0 A short introduction to short read analysis<a class="headerlink" href="#a-short-introduction-to-short-read-analysis" title="Permalink to this headline">¶</a></h1>
<p>Now that we have assembled a genome and have some idea of its content, we can use it as a foundation to quantify biological information. We can get <em>billions</em> of sequences from a single Illumina sequencing run, and we can use these sequences to gain insights into biological processes. We can align reads to a reference genome and identify base pair differences between different plants or cultivars. We can align sequenced mRNAs to gene models or a genome to measure gene expression (RNAseq). We can also look at epigenetic changes such as methylation (bisultife-seq), chromatin modifications (ChIPseq), or chromatin accessibility and conformation (ATACseq, HiC). There are many different ‘-seqs’, and maybe someday you can create <em>your own</em> ‘-seq’ to measure increasingly complex things. At the base of analyzing all of these datasets is aligning the raw data to a reference genome. We can do this the ‘old fashioned’ way (circa 2010) where we align the reads with base pair precision to the genome and store this information, or we can quasi or pseduo align and record <em>roughly</em> where the read maps to. There are pros and cons to both approaches, which we will discuss below.
Today we will go over how to access short read Illumina data, how to align/map RNAseq data, and a basic way to assess sample distribution and sample quality using a principle component analysis. We will continue this analysis in the next class where you will run alignment on a large dataset and we will run differential gene expression and network co-expression analyses. Here, we will use one set of tools that run comparatively quickly, but we will use a slightly different, though conceptually similar pipeline in the next activity.</p>
<p>The basic principle of RNA sequencing (RNAseq) analysis is to use the number of reads that align to different genes in the genome as a proxy of how much those genes are expressed. RNAseq libraries are made from random messenger RNAs, so the pool of sequences is a good representation of what’s being expressed in a given tissue at a particular timepoint. There are different types of RNAseq libraries, but here we will focus on the most common approach where mRNA is enriched using poly T beads, reverse transcribed into DNA, framgented, and ligated to adapters for sequencing. A gene that is highly expressed will be represented many times in the pool of randomly sampled mRNAs. If we look at the alignmnet of RNAseq reads to a highly expressed gene, we can see many reads are aligning (see below). A gene that has low expression will have comparably few reads aligning to it. This is the foundational concept of RNAseq analysis and we can survey gene expression on a genome-wide level by counting how many reads are aligning to each gene model.
<img alt="hi_low_gene_expression.png" src="../../_images/hi_low_gene_expression.png" /></p>
<div class="section" id="plant-biology-learning-objectives">
<h2>Plant Biology Learning Objectives:<a class="headerlink" href="#plant-biology-learning-objectives" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Understand the principles of gene expression analyses and RNAseq</p></li>
<li><p>Explain the fundamentals of Illumina data generation and analysis</p></li>
<li><p>Understand the distinction between mapping and quasi-mapping Illumina data</p></li>
<li><p>Introduce downstream analyses of gene expression data (differential expression and gene co-expression network analyses)</p></li>
<li><p>Introduce the frontiers of gene expression analyses (single cell RNAseq)</p></li>
</ul>
</div>
<div class="section" id="python-learning-objectives">
<h2>Python Learning Objectives:<a class="headerlink" href="#python-learning-objectives" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Understand the principles of dimensionality reduction including feature selection and feature extraction</p></li>
<li><p>Learn how to filter, subset, and transform pandas dataframes</p></li>
<li><p>Learn how to use scikit-learn to perform principal component analysis</p></li>
</ul>
</div>
</div>
<div class="section" id="introduction-to-gene-expression-rnaseq-analysis">
<h1>10.1 Introduction to gene expression (RNAseq) analysis<a class="headerlink" href="#introduction-to-gene-expression-rnaseq-analysis" title="Permalink to this headline">¶</a></h1>
<br>
In this lesson, we will learn about different read alignment and quantification programs, analyze gene expression data from maize, interpret the outputs, and run some quality control analyses on a matrix of RNAseq expression values. In the activity, we will run an RNAseq analysis on data downloaded from the SRA. This is simply a starting point for RNAseq, and there are numerous downstream analyses that can be run to identify genes that are differentially expressed between any two samples/conditions/timepoints, construct gene co-expression networks, compare expression between species, and identify genes or pathways that are important for a given biological process. Below is a broad introduction to RNAseq analysis: <div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">YouTubeVideo</span>

<span class="n">YouTubeVideo</span><span class="p">(</span><span class="s2">&quot;Os5BuF4T6T4&quot;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="mi">960</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="mi">540</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<iframe
    width="960"
    height="540"
    src="https://www.youtube.com/embed/Os5BuF4T6T4"
    frameborder="0"
    allowfullscreen
></iframe>
</div></div>
</div>
</div>
<div class="section" id="downloading-data-from-the-ncbi-sequence-read-archive-sra">
<h1>Downloading data from the NCBI Sequence Read Archive (SRA)<a class="headerlink" href="#downloading-data-from-the-ncbi-sequence-read-archive-sra" title="Permalink to this headline">¶</a></h1>
<p><img alt="ncbi1.jpeg" src="../../_images/ncbi1.jpeg" />
<br>
The National Center for Biotechnology Information (NCBI) maintains a massive repository, the Sequence Read Archive (SRA) for next generation and third generation sequencing data: <br>
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">https://www.ncbi.nlm.nih.gov/sra</a> <br>
The SRA has petabytes ( &gt;1,000,000,000,000,000 base pairs) of raw data from many living and some extinct species on Earth.  You can search for your favorite species in the search bar at the top and see if there are any sequences available. Some of the SRA data has been extensively mined, but most of it has only been analyzed <em>superficially</em>. There are over 30,000 RNAseq samples for Arabidopsis and ~16,000 for maize on the SRA from all kinds of tissues, stresses, developmental time points, etc. Some of this data is defintley low quality, but most was probably carefully collected. What additional insights could be gained by analyzing ALL THE DATA using a common framework? What can be discovered from reanalyzing comparable datasets from many species at once? The SRA is a rich treasure trove of data, and it’s free to use. Most journals require that sequencing data must be deposited in a public repository like the SRA, so keep this in mind. We can download data from the SRA directly to hpcc. <br> <br>
<strong>First,</strong> search the sra database for your favorite species. You can try the common name, but you will probably only find data using its scientific name. <br>
If click on an entry from your search, you will see something like this:
<img alt="ncbi2.jpeg" src="../../_images/ncbi2.jpeg" />
There is a lot of information here, but I have highlighted a few key details. This sample is RNAseq from the orphan grain crop tef under watterlogging (flooding) conditions. It also lists who collected the data, the sequencing technology, and whether the data is single or paired end (as discussed below). At the bottom, I have highlighted a number beginning with <code class="docutils literal notranslate"><span class="pre">SRR</span></code>. This is the unique sequence read identifier for this sample, and we can download data using this identifier.<br>
Run a search for your favorite species and answer a few questions. Record the SRA number, and we will download this data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## What species did you choose? </span>


<span class="c1">## How many samples are there for your species? </span>


<span class="c1">## For one entry, provide the SRR identifier, the type of data, and the tissue it was collected from. </span>
</pre></div>
</div>
</div>
</div>
<p>What is the format for short read sequencing data? We are familiar with the .fasta format of sequence data, and short read Illumina data has a similar format that builds in quality scores.
<br>
Most Illumina data you will encounter is stored in a .fastq format as shown below: <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@SRR1620941.1 AMELIA:320:C2J1FACXX:3:1101:1504:2233 length=202
CTGGGATGTTTGGGCGGCTTCAGATCTGTGGAGTTCGACTCCCTGCATGAGCACCCGGGCTCCTCGCTTGAGAGGCGATC
GTGGTCGCTCACAGCCTCGCCCGACGCCTCCGCTGCCGAGGGATGGGTGGACGAGCTCCGGCGGCTGCGCGTCGCCGAGC
TCCGTCGCGACGTCGAGCGATGCGATCTCTCTATCGGGACGC
+SRR1620941.1 AMELIA:320:C2J1FACXX:3:1101:1504:2233 length=202
B@@FFDDFHHHFHJJJJJJJJJJIJJJIJJJJJHHGIGGIIIJHHHHHFFFFFEDDDDCDDDDCDDDDDDBCDDD@&lt;9@D
BCBDDDDBDDDDDCBBCDBD&lt;+=BDFFFFHHHHHJJJJIJJIGGHJJGIJIJJGJHFEHFFCDDDDDDD8BDDBDDDBBB
DCCDABDDDDDDBDD@BBB&gt;ABDDBDD@CDCCCCCBBD&lt;BB&gt;
@SRR1620941.2 AMELIA:320:C2J1FACXX:3:1101:2085:2106 length=202
CCTGCTTCGCTGCCTGGACTGCCAGCGCCGCTGCTTCCTCTGCAGCCTTCAGGCGTGCAGCCCGGCTGGGGAAGTCGCCC
AGCTCTCCTTCGCTGGTTCCCCTTCCCTACACATGTCGACCTCTGCTGAGCGCGCTGCCAGGCTCAAGGCCAAGGGAACC
AGCGAAGGAGAGCTGGGCGACTTCCCCAGCCGGGCTGCACGC
+SRR1620941.2 AMELIA:320:C2J1FACXX:3:1101:2085:2106 length=202
CCCFFFFFHHHHHIGIJGIIJJJJJJJJGIJJJIIJHJJJJJJJJIBGIEA?EADFDDDDEDBDBBDDDDDDDDCC@5?@
BB?CCDDDCCDBDDD&gt;A@CC@+=BFFFFFHHHHHJJJJIJJIJJIIJJJJJIIIIIJIIJIJIIJIEHIJHHHHFFFFDD
CDDDDDDDBDBDDDDDDDD9BBDDADDDBDBBBBDDDDDDDD
</pre></div>
</div>
 <br>
1. The first line (starting with @SRR...) is a unique identifier for that read. Remember, experiments can have billions of reads, so the names/identifiers can be quite long. <br>
2. The second line (starting with ATG...) is the actual sequence. This is the bit we care most about. <br>
3. The third line (starting with +SRR...) is the same unique identifier for that read. <br>
4. The fourth line (seemingly random characters) contains the quality scores. Each letter, symbol, or number corresponds to a sequence position. The sequences and quality scores will always have the same number of characters. During sequencing, the Illumina machine will asign a quality score for each base pair based on how confident the base call was. Some clusters that are close together can have mixed signals and as the Illiumina run progresses, reagents wear down and quality scores tend to be lower. We can use these quality scores to remove 'bad' bases and during read alignment. </div>
<div class="section" id="challenge-problem-download-sra-data-onto-hpcc">
<h1>Challenge problem: Download SRA data onto HPCC:<a class="headerlink" href="#challenge-problem-download-sra-data-onto-hpcc" title="Permalink to this headline">¶</a></h1>
<br>
Lets say you want to reanalyze this data or combine data from several experiments. We can use the SRA-Toolkit program to download data from the SRA. The SRA toolkit is installed on HPCC and can be accessed by loading the module from a **developmental node** or in a SLURM submission script. Simply type: <br> <br>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">module</span> <span class="pre">load</span> <span class="pre">SRA-Toolkit/2.10.7-centos_linux64</span></code></p>
<p><br> <br>
We can then download the data using the following command: <br> <br>
<code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">fastq-dump</span> <span class="pre">SRR1620942</span></code> <br>
<br>
<strong>Note:</strong> I have included an example SRR number here, but you should download the file you described above for your favorite species! Also, you need to specify the output directory (<code class="docutils literal notranslate"><span class="pre">--outdir</span> <span class="pre">&lt;path&gt;</span></code>) or the reads will be downloaded by default into your present working directory. The SRR number can be found at the bottom left hand corner of the info page for that sample.
<br>
<br>
Because fastq files can get quite large, raw sequences should be downloaded into scratch (<code class="docutils literal notranslate"><span class="pre">/mnt/scratch/MSU_ID/</span></code>) or your home directory (<code class="docutils literal notranslate"><span class="pre">/home/MSU_ID</span></code>). Remember, files on scratch are deleted automatically after 45 days, so don’t leave anything important there! I have also downloaded some test data into the course directory
<br>
Some data is single end and some will be paired end. Single end reads are just sequenced in one direction where paired end data is sequenced in both directions where one read corresponds to one end of a sequenced fragment and its pair corresponds to the other end. If the data is paired end, you can split it into Read 1 and Read 2 by using the split flag as follows: <br> <br>
<code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">fastq-dump</span> <span class="pre">--split-3</span> <span class="pre">SRR1620942</span></code> <br> <br>
It is also useful to rename the files from the default SRR number into something more informative (i.e. species name, tissue type, sample, treatment, etc).</p>
<p>This approach is fine when downloading a few files, but what if we want to download <em>a lot</em> of .fastq files? We could recursively run each sample one at a time, but that would take ages and be a poor use of our time. Instead, we can write a simple shell script and list to download a batch of files as follows. Note, shell script is another programming language with different rules or syntax compared to Python. For our purposes, we will only need to know a few simple functions such as loops, and here is a nice tutorial on the basics of shell programming: <a class="reference external" href="https://www.shellscript.sh/">https://www.shellscript.sh/</a>. Below is an example script we can use to download sequencing data in bulk: <br></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#!/usr/bin/bash
fastq-dump $1
</pre></div>
</div>
<br>
Where $1 sets the first argument in command line.  
<br>
Then, we can create a list of SRA identifiers for the files we want to download with one identifier per line as follows: <br>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>SRR1620941
SRR940286
SRR940284
SRR940283
SRR940276
SRR940278
SRR940279
SRR1620975
SRR1620976
SRR1620977
</pre></div>
</div>
<br>
We can then read this list and feed it into our shell script to download the files recursively `|` connects the 
    $cat list_SRA.txt | xargs -n 1 bash get_SRR_data.sh</div>
<div class="section" id="dimensionality-reduction">
<h1>10.2 Dimensionality reduction<a class="headerlink" href="#dimensionality-reduction" title="Permalink to this headline">¶</a></h1>
<br>
Large-scale datasets are difficult to interpret 'by eye'. We can graph simple metrics such as leaf number, photosynthetic rate, number of caterpillars on a plant, etc. and run a simple statistical test to see if two samples differ from expectatiobs. With gene expression analyses, we have a data matrix with tens of thousands of genes across any number of samples or experimental conditions. How can we simplify this dataset so it can be easily visualized or used in downstream analyses such as predictive modeling or machine learning? <br>
Dimensionality reduction is an approach to transform complex, high dimensionality data (such as an RNAseq matrix) into a low dimensional space that still captures the biologically meaningful patterns of the experiment. We could use principle component analysis (PCA) to transform a large expression matrix into two or three meaningful principle components to see the relation of samples over the experimental space. The following video introduces the concept of dimensionality reduction and how to run a PCA on an expression matrix using the powerful python package `scikit-learn`.<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">YouTubeVideo</span>

<span class="n">YouTubeVideo</span><span class="p">(</span><span class="s2">&quot;Rq4ZoDutfqU&quot;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="mi">960</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="mi">540</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<iframe
    width="960"
    height="540"
    src="https://www.youtube.com/embed/Rq4ZoDutfqU"
    frameborder="0"
    allowfullscreen
></iframe>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import modules</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load the test dataset</span>
<span class="n">expr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Lesson10_tef_raw_TPM.txt&#39;</span><span class="p">)</span>
<span class="n">expr_df</span> <span class="o">=</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">expr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">expr_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="n">expr_df</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="n">ipykernel_30886</span><span class="o">/</span><span class="mf">2341020248.</span><span class="n">py</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> 
<span class="g g-Whitespace">      </span><span class="mi">6</span> <span class="c1"># Load the test dataset</span>
<span class="ne">----&gt; </span><span class="mi">7</span> <span class="n">expr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Lesson10_tef_raw_TPM.txt&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">8</span> <span class="n">expr_df</span> <span class="o">=</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">expr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="g g-Whitespace">      </span><span class="mi">9</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/util/_decorators.py</span> in <span class="ni">wrapper</span><span class="nt">(*args, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">309</span>                     <span class="n">stacklevel</span><span class="o">=</span><span class="n">stacklevel</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">310</span>                 <span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">311</span>             <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">312</span> 
<span class="g g-Whitespace">    </span><span class="mi">313</span>         <span class="k">return</span> <span class="n">wrapper</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">read_csv</span><span class="nt">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">584</span>     <span class="n">kwds</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds_defaults</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">585</span> 
<span class="ne">--&gt; </span><span class="mi">586</span>     <span class="k">return</span> <span class="n">_read</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">587</span> 
<span class="g g-Whitespace">    </span><span class="mi">588</span> 

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">_read</span><span class="nt">(filepath_or_buffer, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">480</span> 
<span class="g g-Whitespace">    </span><span class="mi">481</span>     <span class="c1"># Create the parser.</span>
<span class="ne">--&gt; </span><span class="mi">482</span>     <span class="n">parser</span> <span class="o">=</span> <span class="n">TextFileReader</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">483</span> 
<span class="g g-Whitespace">    </span><span class="mi">484</span>     <span class="k">if</span> <span class="n">chunksize</span> <span class="ow">or</span> <span class="n">iterator</span><span class="p">:</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">__init__</span><span class="nt">(self, f, engine, **kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">809</span>             <span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwds</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">    </span><span class="mi">810</span> 
<span class="ne">--&gt; </span><span class="mi">811</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_engine</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_engine</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">engine</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">812</span> 
<span class="g g-Whitespace">    </span><span class="mi">813</span>     <span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/readers.py</span> in <span class="ni">_make_engine</span><span class="nt">(self, engine)</span>
<span class="g g-Whitespace">   </span><span class="mi">1038</span>             <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1039</span>         <span class="c1"># error: Too many arguments for &quot;ParserBase&quot;</span>
<span class="ne">-&gt; </span><span class="mi">1040</span>         <span class="k">return</span> <span class="n">mapping</span><span class="p">[</span><span class="n">engine</span><span class="p">](</span><span class="bp">self</span><span class="o">.</span><span class="n">f</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">)</span>  <span class="c1"># type: ignore[call-arg]</span>
<span class="g g-Whitespace">   </span><span class="mi">1041</span> 
<span class="g g-Whitespace">   </span><span class="mi">1042</span>     <span class="k">def</span> <span class="nf">_failover_to_python</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/c_parser_wrapper.py</span> in <span class="ni">__init__</span><span class="nt">(self, src, **kwds)</span>
<span class="g g-Whitespace">     </span><span class="mi">49</span> 
<span class="g g-Whitespace">     </span><span class="mi">50</span>         <span class="c1"># open handles</span>
<span class="ne">---&gt; </span><span class="mi">51</span>         <span class="bp">self</span><span class="o">.</span><span class="n">_open_handles</span><span class="p">(</span><span class="n">src</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">52</span>         <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
<span class="g g-Whitespace">     </span><span class="mi">53</span> 

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/parsers/base_parser.py</span> in <span class="ni">_open_handles</span><span class="nt">(self, src, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">220</span>         <span class="n">Let</span> <span class="n">the</span> <span class="n">readers</span> <span class="nb">open</span> <span class="n">IOHandles</span> <span class="n">after</span> <span class="n">they</span> <span class="n">are</span> <span class="n">done</span> <span class="k">with</span> <span class="n">their</span> <span class="n">potential</span> <span class="n">raises</span><span class="o">.</span>
<span class="g g-Whitespace">    </span><span class="mi">221</span>         <span class="s2">&quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">222</span><span class="s2">         self.handles = get_handle(</span>
<span class="g g-Whitespace">    </span><span class="mi">223</span><span class="s2">             src,</span>
<span class="g g-Whitespace">    </span><span class="mi">224</span><span class="s2">             &quot;r&quot;,</span>

<span class="nn">~/miniconda3/envs/jbook/lib/python3.8/site-packages/pandas/io/common.py</span> in <span class="ni">get_handle</span><span class="nt">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">699</span><span class="s2">         if ioargs.encoding and &quot;b&quot; not in ioargs.mode:</span>
<span class="g g-Whitespace">    </span><span class="mi">700</span><span class="s2">             # Encoding</span>
<span class="ne">--&gt; </span><span class="mi">701</span><span class="s2">             handle = open(</span>
<span class="g g-Whitespace">    </span><span class="mi">702</span><span class="s2">                 handle,</span>
<span class="g g-Whitespace">    </span><span class="mi">703</span><span class="s2">                 ioargs.mode,</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] No such file or directory: &#39;Lesson10_tef_raw_TPM.txt&#39;
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "PlantsAndPython/PlantsAndPython",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./Q_10_RNASEQ_ANALYSIS/0_Lessons"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="../0_Overview.html" title="previous page">Welcome / Bienvenid&#64;</a>
    <a class='right-next' id="next-link" href="../1_Practice.html" title="next page">Practice: Introduction to gene expression (RNAseq) analysis</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By  <br/>
        
            &copy; Copyright By the authors.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>