{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <p style=\"color:#2C6A7F\"> 02. Unix Commands for Data Mining\n",
    "## <p style=\"color:#2C6A7F\"> Practice\n",
    "\n",
    "***Author:  Dr. Alejandra Rougon*** \n",
    "\n",
    "<p style=\"color:#9D994B\"><a rel=\"license\" href=\"http://creativecommons.org/licenses/by-nc/4.0/\"><img alt=\"Creative Commons License\" style=\"border-width:0\" src=\"https://i.creativecommons.org/l/by-nc/4.0/88x31.png\" /></a><br />This work is licensed under a <a rel=\"license\" href=\"http://creativecommons.org/licenses/by-nc/4.0/\">Creative Commons Attribution-NonCommercial 4.0 International License</a>.\n",
    "\n",
    "\n",
    "\n",
    "## <p style=\"color:#7f2c40\">ðŸš´ Exercise 1 \n",
    "Now is time for you to practice what you have learned. Try to solve as many questions as possible. There are usually various ways to solve the problems. \n",
    "\n",
    "\n",
    "Now let's do some data mining in some tiny files. Once you have learn to do this, you will be able to work with large genomic data files. \n",
    "\n",
    "1. On your home directory create a new directory called `Exercise1`.\n",
    "2. Create a new file inside the folder `Exercise1` with `vim` called `ToyPlant.fasta`\n",
    "\n",
    "\tyou can copy/paste the following contents\n",
    "\t\n",
    "\t```\n",
    "\t>Plant_1\n",
    "\tACCACCGATACATGCGGTGCGTTGT\n",
    "\t>Plant_3\n",
    "\tCCACTGTGTTCGAGTTGTGATACAG\n",
    "\t>Plant_3\n",
    "\tCCACTGTGTTCGAGTTGTGATACAG\n",
    "\t>Plant_2\n",
    "\tCCAGCATTTGTAGTCACAACGCCGC\n",
    "\t>Plant_4\n",
    "\tTAGAGTTGTACACGCGTTTGTACGA\n",
    "\t>Plant_4\n",
    "\tTAGAGTTGTACACGCGTTTGTACGA\n",
    "\t>Plant_1 \n",
    "\tACCACCGATACATGCGGTGCGTTGT\n",
    "\t```\n",
    "3. See the file permissions\n",
    "3. Give permissions of writing, reading and executing to everyone in ToyPlants.fasta\n",
    "4. How many lines does the file have?\n",
    "5. How many records does the file have?\n",
    "6. How many unique records does the file have?\n",
    "7. Calculate the total amount of bases [the genome size]\n",
    "8. How many sequences contain the string `GATACA` [Specific sequence strings that may have a particular function or structure are called **motives** or **domains**.]\n",
    "9. Make a backup of that file in `Documents`.\n",
    "10. In the folder `Exercise1` create the following file called ToyPlant.genes\n",
    "\t\t\n",
    "\t```\n",
    "\tchr1 height ht-1 100 1000 + (100-150,400-500,900-1000)\n",
    "\tchr1 height ht-2 100 1000 + (100-150,900-1000)\n",
    "\tchr1 resist res-1 1500 2000 + (1500-1750,1800-1850,1099-2000)\n",
    "\tchr1 resist res-2 1500 2000 + (1500-2000)\n",
    "\tchr2 color color-1 3400 4200 - (3400-3600,4000-4200)\n",
    "\tchr2 color color-2 3400 4200 - (3400-3550,3800-3900,4000-4200)\n",
    "\tchr2 color color-3 3400 4200 - (3400-3600,3800-3900,4100-4200)\n",
    "\tchr3 fruit fru-1 50 800 + (50-400,700-800)\n",
    "\tchr3 fruit fru-1 1100 1500 + (1100-1200,1450-1500)\n",
    "\tchr3 smell smell-1 2000 2600 - (2000-2300,2500-2600)\n",
    "\tchr3 smell smell-2 2000 2600 - (2000-2050,2200-2300,2500-2600)\n",
    "\tchr4 dev dev-1 3100 3700 - (3100-3500,3600-3700)\n",
    "\tchr4 dev dev-2 3100 3700 - (3100-3200,3400-3500,3600-3700)\n",
    "\tchr4 height2 ht2-1 4500 4800 + (4500-4800)\n",
    "\tchr5 shape shape9-1 200 1000 - (200-450,550-650,800-1000)\n",
    "\tchr5 shape shape10-1 110 1700 + (110-1400,1500-1700)\n",
    "\t```\n",
    "\n",
    "11. How many transcripts does the file show? (all lines)\n",
    "12. How many different chromosomes does the file have? (column 1; the file separator is a space)\n",
    "13. How many different genes does the genome have? (column 2)\n",
    "\n",
    "\n",
    "### <p style=\"color:#7f2c40\">ðŸš´ Exercise 2\n",
    "We are studying some proteins that are involved in pathogenicity called effectors found on the sequences of the phytopathogen *Hyaloperonospora arabidopsidis*. We want to know how many of those sequences are **RxLR** effectors (RxLR for containing Arginine, any amino acid, Leucine and Arginine). We also want to know which ones are cysteine-rich. And which of the RxLR effectors belong to a specific strain called Emoy2. \n",
    "\n",
    "At the moment, we are only going to look for the strings 'RxLR' and 'cysteine-rich' within the decription line. However, you could look for specific domains within the sequences using other tools to eliminate the line breaks and to look for ambiguous bases or amino acids. To convert the fasta file into a tabular file with each fasta record  in a single line, you can use this command `awk -v RS='\\n>' -v ORS='\\n>' -v OFS='' -F'\\n' '{$1=$1 \"\\t\"}1'  file.fasta`. Then you can select the second column and look for a string. You can use a dot `.` to find ambiguos bases or amino acids. The `.` is a regular expression that represents any character. So, instead of `RxLR`, if you look into the actual sequence, you would have to use `R.LR`. In order to search for regular expressions with `grep` you have to use the option `-E`. If you want to see the string colored in your search with `grep`, use the option `--color`.\n",
    "\n",
    "1. Go to the folder `Exercise1` and create a new directory called `Analysis`\n",
    "2. Upload the following file to your [virtual terminal](https://bit.ly/3d9BRCG) <br>[Hp1.fasta](https://drive.google.com/file/d/1JywPKMg_mtBeMPDYh_OORinlZc5VjTpj/view?usp=sharing) <br> \n",
    "3. How many records does `Hp1.fasta` have?\n",
    "\t\n",
    "\t For the next three questions we will analyze the identifiers and not the sequences\n",
    "\n",
    "4. How many of those records are **RxLR** proteins?\n",
    "5. How many of those records are **cysteine-rich** proteins?\n",
    "6. How many of the **RxLR** proteins belong to the strain **Emoy2**?\n",
    "\n",
    "-----------\n",
    "\n",
    "Thank you for completing this activity!\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
